{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Process-CCSs\" data-toc-modified-id=\"Process-CCSs-1\">Process CCSs</a></span><ul class=\"toc-item\"><li><span><a href=\"#Setup\" data-toc-modified-id=\"Setup-1.1\">Setup</a></span></li><li><span><a href=\"#PacBio-amplicons\" data-toc-modified-id=\"PacBio-amplicons-1.2\">PacBio amplicons</a></span></li><li><span><a href=\"#CCS-stats-for-PacBio-runs\" data-toc-modified-id=\"CCS-stats-for-PacBio-runs-1.3\">CCS stats for PacBio runs</a></span></li><li><span><a href=\"#Align-CCSs-to-amplicons\" data-toc-modified-id=\"Align-CCSs-to-amplicons-1.4\">Align CCSs to amplicons</a></span></li><li><span><a href=\"#Write-valid-CCSs\" data-toc-modified-id=\"Write-valid-CCSs-1.5\">Write valid CCSs</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This Python Jupyter notebook processes the PacBio circular consensus sequences (CCSs) to extract barcodes and call mutations in the gene. It then builds consensus sequences for barcoded variants from the mutations, to build a barcode-variant lookup table."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process CCSs\n",
    "First, process the PacBio CCSs to extract barcodes and call mutations in the gene."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define the background for CCS mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "background = \"MERS\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "\n",
    "Import Python modules\n",
    "\n",
    "Plotting is done with [plotnine](https://plotnine.readthedocs.io/en/stable/), which uses ggplot2-like syntax.\n",
    "\n",
    "The analysis uses the Bloom lab's [alignparse](https://jbloomlab.github.io/alignparse) and [dms_variants](https://jbloomlab.github.io/dms_variants) packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import collections\n",
    "import math\n",
    "import os\n",
    "import re\n",
    "import time\n",
    "import warnings\n",
    "\n",
    "import alignparse\n",
    "import alignparse.ccs\n",
    "from alignparse.constants import CBPALETTE\n",
    "import alignparse.minimap2\n",
    "import alignparse.targets\n",
    "import alignparse.consensus\n",
    "import alignparse.utils\n",
    "\n",
    "import dms_variants\n",
    "import dms_variants.codonvarianttable\n",
    "import dms_variants.plotnine_themes\n",
    "import dms_variants.utils\n",
    "\n",
    "from IPython.display import display, HTML\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from plotnine import *\n",
    "\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set [plotnine](https://plotnine.readthedocs.io/en/stable/) theme to the one defined in [dms_variants](https://jbloomlab.github.io/dms_variants):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "theme_set(dms_variants.plotnine_themes.theme_graygrid())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Versions of key software:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using alignparse version 0.6.0\n",
      "Using dms_variants version 1.4.3\n",
      "Using pandas version 1.3.5\n"
     ]
    }
   ],
   "source": [
    "print(f\"Using alignparse version {alignparse.__version__}\")\n",
    "print(f\"Using dms_variants version {dms_variants.__version__}\")\n",
    "print(f\"Using pandas version {pd.__version__}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ignore warnings that clutter output:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.simplefilter('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the configuration file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('config.yaml') as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make output directory for figures:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(config['process_ccs_dir'], exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PacBio reads\n",
    "Get the amplicons sequenced by PacBio as the alignment target along with the specs on how to parse the features:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>library</th>\n",
       "      <th>bg</th>\n",
       "      <th>run</th>\n",
       "      <th>name</th>\n",
       "      <th>fastq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>MERS</td>\n",
       "      <td>230124</td>\n",
       "      <td>lib51_MERS_230124</td>\n",
       "      <td>results/ccs/lib51_MERS_230124_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib52</td>\n",
       "      <td>MERS</td>\n",
       "      <td>230124</td>\n",
       "      <td>lib52_MERS_230124</td>\n",
       "      <td>results/ccs/lib52_MERS_230124_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pacbio_runs = (\n",
    "    pd.read_csv(config['pacbio_runs'], dtype=str)\n",
    "    .drop(columns=['ccs'])\n",
    "    .assign(name=lambda x: x['library'] + '_' + x['bg'] + '_' + x['run'],\n",
    "            fastq=lambda x: config['ccs_dir'] + '/' + x['name'] + '_ccs.fastq.gz'\n",
    "            )\n",
    "    )\n",
    " \n",
    "pacbio_runs.drop(pacbio_runs[pacbio_runs['bg'] != background].index,inplace=True)\n",
    "\n",
    "display(HTML(pacbio_runs.to_html(index=False)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PacBio amplicons\n",
    "Get the amplicons sequenced by PacBio as the alignment target along with the specs on how to parse the features:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading amplicons from data/PacBio_amplicon_MERS.gb\n",
      "Reading feature parse specs from data/feature_parse_specs_MERS.yaml\n"
     ]
    }
   ],
   "source": [
    "print(f\"Reading amplicons from {config['amplicons_' + background]}\")\n",
    "print(f\"Reading feature parse specs from {config['feature_parse_specs_' + background]}\")\n",
    "\n",
    "targets = alignparse.targets.Targets(\n",
    "                seqsfile=config['amplicons_' + background],\n",
    "                feature_parse_specs=config['feature_parse_specs_' + background])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Draw the target amplicons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving plot to results/process_ccs/amplicons_MERS.pdf\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAmkAAADaCAYAAAD9sNUgAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA2KElEQVR4nO3dd3gUVd/G8e+mF9ILpAkkhCo9CAjSFQQRFASxURTEioj4oPDY9cECr11B6SAiKCCogCK9hi4gnURCAoEQQnrbff+ILKybhJ7dmPtzXXtd7Mw5s78dx+TOzJwzBpPJZEJERERE7IqDrQsQEREREWsKaSIiIiJ2SCFNRERExA4ppImIiIjYIYU0ERERETukkCYiIiJihxTSREREROyQQpqIiIiIHVJIExEREbFDCmkiUiamTp2KwWDAYDCwcuVKq/Umk4kaNWpgMBho166defn5PsW9BgwYYG732muvWaxzdnbmpptuYvDgwZw4ccLq8zIzM3n33Xdp2LAh3t7eeHl5ERUVRZ8+fVi1atUN2AMiIlfGydYFiEjF4uXlxaRJkyyCGMCqVas4fPgwXl5eVn169+7NiBEjrJYHBQVZLVuyZAk+Pj5kZGSwbNkyxo0bx/r169mxYwfOzs4AFBYWcscdd/DHH38wcuRIbrnlFgAOHjzIokWLWLNmDW3btr0O31ZE5OoppIlImerbty+zZs3is88+w9vb27x80qRJtGzZknPnzln1qVy5Mi1atLis7Tdt2pTAwEAAOnXqxOnTp5kyZQpr166lffv2AKxevZr169czefJkBg4caO7buXNnnn76aYxG47V8RRGR60KXO0WkTPXr1w+A2bNnm5elpaXx/fffM2jQoOv+eTExMQCcPHnSvCwlJQWAkJCQYvs4OOhHo4jYnn4SiUiZ8vb2pnfv3kyePNm8bPbs2Tg4ONC3b99i+5hMJgoKCqxeJpPpkp939OhRAGrWrGleFhMTg7OzM8OGDWPWrFkkJSVd47cSEbn+FNJEpMwNGjSIzZs3s2fPHgAmT57MfffdV+z9aACff/45zs7OVq9Zs2ZZtS0sLKSgoICzZ88yd+5cvvjiC/r160eTJk3MbapVq8aXX35JYmIiDz30EKGhoYSGhtK/f3/WrFlzY760iMgV0j1pIlLm2rZtS1RUFJMnT2bAgAHExsYybty4Etv36dOHkSNHWi2PjIy0WlalShWL923atGHatGlW7QYNGkSvXr34+eef2bhxIxs3bmTmzJnMmDGDd999t9jPExEpSwppIlLmDAYDAwcO5OOPPyYnJ4eaNWty2223ldg+KCjIfG/Zpfz222/4+Phw5swZJk6cyPfff88zzzzDl19+adXWx8eHfv36me+T27NnD506dWL06NEMHjwYX1/fq/p+IiLXgy53iohNDBgwgNOnT/Pll19ajLC8Vg0bNiQmJoY77riDuXPncvvttzNx4kRiY2Mv2bdevXrcf//95Ofnc+DAgetWk4jI1VBIExGbCAsLY+TIkXTv3p3+/fvfkM8wGAx89tlnODo6MmbMGPPylJQU8vLyiu2zb98+AEJDQ29ITSIil0uXO0XEZsaOHXtZ7U6ePMnGjRutlnt7e1O3bt1S+0ZHRzNkyBA+//xz1q5dS+vWrVmxYgXDhg3jwQcf5NZbbyUgIIDk5GRmz57NkiVLeOSRRwgPD7+q7yQicr0opImI3Zs3bx7z5s2zWt6qVSvWrl17yf6vvvoq06dP55VXXuH333+nRYsWDBo0iBUrVjBjxgxOnz6Nu7s7devW5ZNPPuGJJ564EV9DROSKGEyXM9GQiIiIiJQp3ZMmIiIiYocU0kRERETskEKaiIiIiB1SSBMRERGxQwppIiIiInZIIU1ERETEDimkiYiIiNghhTQRERERO6SQJiIiImKHFNJERERE7JBCmoiIiIgdUkgTERERsUMKaSIiIiJ2SCFNRERExA4ppImIiIjYIYU0ERERETukkCYiIiJihxTSREREROyQQpqIiIiIHVJIExEREbFDCmkiIiIidkghTURERMQOKaSJiIiI2CGFNBERERE7pJAmIiIiYocU0kRERETskEKaiIiIiB1SSBMRERGxQwppIiIiInZIIU1ERETEDimkiYiIiNghhTQRERERO6SQJiIiImKHFNJERERE7JBCmoiIiIgdUkgTERERsUMKaSIiIiJ2SCFNRERExA4ppImIiIjYIYU0ERERETukkCYiIiJihxTSREREROyQQpqIiIiIHVJIExEREbFDCmkiIiIidkghTURERMQOKaSJiIiI2CGFNBERERE7pJAmIiIiYocU0kRERETskEKaiIiIiB1SSBMRERGxQwppIiIiInZIIU1ERETEDimkiYiIiNghhTQRERERO6SQJiIiImKHFNJERERE7JBCmoiIiIgdUkgTERERsUMKaSIiIiJ2SCFNRERExA4ppImIiIjYIYU0ERERETukkCYiIiJihxTSREREROyQQpqIiIiIHVJIExEREbFDCmkiIiIidkghTURERMQOKaSJiIiI2CGFNBERERE7pJAmIiIiYoecbF2AiIiI/PukpaVx8OBBMjMzMZlMti7HJpycnAgODiYqKgpHR8cr738DahIREZEKKi8vj3nz5nHw4EEcHBxwdXfHwaEiXrgzkZ+XT15uLq6urnTr1o369etf0RYU0kREROS6MJlMzJ49m+OJibTt0ZVqdWrh4upq67JsxmQykZp8iu1r1jN//nxcXV2pWbPmZfeviNFWREREboCkpCTi4uJod89d1GzUoEIHNACDwYB/5WDa33s3lSPC2LBxwxX1V0gTERGR62Lfvn24ubtzU80ati7Frjg4OFCjfj3i4+LJycm5/H43sCYRERGpQNLT0/H296ug96CVzjcoAJPJREZGxmX30V4UERGR68JoNOLgWHy0eG7AY3S4uUkZV3R9nDl9mjCDG3OmTr/qbTg6Fg0DMBqNl91HIU1ERETEDimkiYiISLl2Jfd5lScKaSIiIlJmfv9lKR1ubkKkmw9dmrZk68ZN5nVzp8+kZ+v21PMPoa5fFXq3u53tm2Mt+o977U2iKwWwfXMs3Vu2JdLNhymffA7AwT/38di9fannH0KUhx+dGjZjwew55r45OTm8PuI/NA2LpLqrNx3rN2X+N99a1Tjrq0k0r1aTKA8/+nTswtFDh4v9LnOmTqdTgxgi3XxoGhbJ2NGvUFBQcD12E6B50kRERKSMnEw6wctPPsvzr43B18+PT8d+wIOdu7P24G4Cg4NJiIun9yMPUjUqkvy8fOZ/8y292nTi111biKoZbd5Ofl4eTz84gMHDn+Wl/72Bj58fRw4e4u6WbQmNCOeNj8cRVKUy+3fv5fhfx8z9nnlwAL//vISRb71G7ZvrsfDb73j6wQEUFhbS++EHAfh18c+8OOQp+gx4mB7392Hnlq08ef/DVt9lwviPePvFlxk8/FleGTeWg3/u593Rr2IsNPLy2Leuy/5SSBMREZEycfbMGSbMnUXrDu0BaN6mNc0iavDVh5/w0jtvMvyV0ea2RqORNrd3ZGfsVr6bOp2X3nnTvC4/P59R77xB9/t6mZc9/WB/nF1cWLBuBV7e3gC06dTRvH7vrj/4+YcFvP3ZRwx48nEA2nW+nZOJSbz/39fNIe2jt8bS/LZW/N+Ur8xtsjIz+fR/75u3lZGezrhX3+SJF58319Xm9k44Ojnx5gujGDpyOP4BAde8v3S5U0RERMqEt4+POaAB+Pj60qpDO7Zt3AwUXa589J4+NKx8ExGOHlR1rsTh/Qc4cuCQ1bY6du1i8X7t8pV0632POaD90+Y16wC4u29vi+U9+vUhIf4vjh87RmFhIX9s3UaXe3pYtLmr970W77es30hmRgbd7+tFQUGB+dW6QztysrPZv3vPZe6R0ulMmoiIiJQJ/6BAq2WBwUEcOXCQjPR0+t3RjYCgIF4d/y7hVW/C1c2NFx57gtx/DAxw9/DAw9PTYllqSgqVQ0NK/Oyzqak4OTlZneEKrlKlaP2ZVJydnSkoKCAwOMiyxsrBFu/PnD4NQOcmLYr9rMRjCSXWcSUU0kRERKRMnDl12mrZ6eRTBIdUYeuGjSQlHGfa4vnUa9jAvD49LY2Q8DCLPgaDwWo7fgEBnExMKvGzff39KSgoIPXMGfz8/c3Lk0+c+Hu9HwFBQTg5OXE6+ZRljSeTrbYF8PUPcwiNCLf6rIjq1Uqs40rocqeIiIiUiXNpaaz9fYXF+3W/r6RJ82bkZBedLXNxcTGvj12/gWNx8Ze17ds6teenefPJSE8vdv0trW8FYNF331ssXzRnHuFVbyIsIgJHR0fqN2nMkvkLLdosnveDxfuYW1vg7uFBUsJxGsY0tXpdj/vRQGfSREREpIz4+vvzwqNDGfH6f/Hx9eXTsR8A8NhzzwDgWakSLz81jKdHvcCJ44mMe+0tqoSFlbZJs+GvjuG3xb/Qs3UHnnzxeYJDqnBw7z6ys7J48sUR1G1Qn6697uH1518kOyuLWvXqsui7eaxYsoyPpk8yb+fZ0f9hYI/eDB842Dy6c8E3cyw+y9vHhxfeeIW3X3yZpITj3Nq+LQ4OBuKPxLFs4SK++v5b3D08rnl/6UyaiIiIlInKIVV469MP+WzsBzx+3wPk5uQwa+kigipXJqhyZSbM/YaU5FMM6nEfX334KWO//JTqNSIva9uR0TVYuH4lEdWq8vKTwxjQvRezJ00lrOpN5jafzJzCI08MYcIHHzKg+71s3xTLJzOnmEd2Atxx912M/fIT1i5fwaM972PNr8v5bLb146CGjniO8VMmsn7FKh67ty+P3/cgsyZOomGzpjhfdDbwWhhMJpPpumxJREREKrT58+eTfPYM3Qc+ZOtS7E5yQiILJ03niSeeIDg4+NId0Jk0EREREbukkCYiIiJyg5m48guXCmkiIiJyXTg7O5OXk2vrMuzS+f3icgX3qymkiYiIyHURGhrKmeRTZGVk2LoUu5N4JA4PT0+8S3giQnEU0kREROS6qF27Ng4ODmz+bSUal3jBmeRT7N+xizp/75/LpdGdIiIict3s2rWL+fPnE1A5mOp1a+Pl54ODg6OtyypzJkwU5OVx4q8Eju7dj7+fH4888ggeVzB/mkKaiIiIXFeHDx9m27ZtHDx4kPz8fFuXY1MBAQHUrVuXFi1aXFFAA4U0ERERuUFMJhP5+fk2vfTZt29f5syZc+mGN4CTkxOOjld/FlGPhRIREZEbwmAwXNFoxhvBaDTi6upq0xqulgYOiIiIiNghhTQRERERO1Tq5c78/HwOHTrE6dOnKSwsLKua7JarqytVq1YlJCQEg8Fg63JERETKjczMTA4cOEB6ejpGo7HMPjckJISVK1eW2eddDoPBgIeHBzVq1MDPz6/kdiUNHNi4cSO///47+fn5uLu54uRcsW9fM5lM5OTkUVBQgL+/P/fffz9BQUG2LktERMSuGY1GFi1axM6dOwFwd3fD0bHsLuQVFBTg5GRfGcZoNJKdnYvRaKR69er06dMHNzc3q3bFhrRt27axaNEiYhrXoWWz+vj7+ZRJ0fbOaDQS91cSS5dvJDM7lyFDhlzRzMEiIiIVzcKFC9m5cyedOrWg/s3ReHq627oku5CXl8/+A3EsWbKOgIBABg0aZDXRrVWUNZlMrF+/jrq1qtP19lYKaBdxcHAgsloYj9zflby8XHbt2mXrkkREROxWZmamOaC1aN5AAe0iLi7O1L85ml73duL48eP89ddfVm2sQlpKSgopKWdoWL+m7rsqgaenOzWqR7B/3z5blyIiImK3Dhw4AED9m6NtXIn9ql49DG+vSuzfv99qnVVIO3fuHABBAb43vLDyLDDAl3Pp52xdhoiIiN1KT0/H3d1NZ9BKYTAYCAjwMeevi1mFtPMjLhyLeQDogh+X8PmEqde/wqtkcA/jg//78pr7vfbWOAzuYVavL7+aXuI2nJwcynR0ioiISHljNBpLHCSwYOEvfPHF1LItqBSOTiGMG/fFNfebOWset97ajcCgOnh4VqNO3da89dZ4cnNzS9mGY7GZ4oqGOyxYtIQt23bx5OMDrqTbDbNh5Y9UvSn8uvRzd3fj91++s1gWWb3qNdUnIiIixVu4cAlbt+7kiScG2LoUANatXUzVqleeKf7ZL/XMWe68syMvvTSMSpU82bx5G2+8OZ5jCYlM+PKDK9q2zcakZmdn4+5+bac/WzRvet36OTg4XPX2RERExHauS6ZocZWZ4h/9nnnmMYv37du3Ij09gw8/msjnn717Rc/yvOyJSgYMfo5pM+eyZ+9+8+XAAYOfA2DDxi106HIfngE18Klcmwf6P0Vy8mlz37j4Yxjcw5g6Yw6DnxxJQFg9mrXuBhRdenz3g88YNeYdgiLq41ulDi+MegOTycTyFWto1Px2KgVG06HLfRw7dtyipn9etmx3R2/uuvcR5n6/iFoNbjP3O3wkrtR+IiIiUnYGDhrG9OnfsWfPfhydQnB0CmHgoGEAbNiwhU6deuPlHYmff00efOhJy0wRdwxHpxCmTpvDkMdHEBRcl+Yt7gSKLj2+996nvPTS21SuUg//gFqMHPl6UaZYvoYmTTvh7RNFp069rTLFPy9bduhwL93vfpi58xZRp25rc7/Dh+NK7VecgAA/8vMLrvg2qcs+k/bfl57j1OkU9u0/zKwpnwAQFBTAho1baNf5Prp27sCcGV+QmZnFmNff4+7eA9i4erHFNl56ZSzdu97O7GmfWzzB4NMvp9ChXStmTv6ETbHbefXNDygsLGT5ynWMfvEZXFxceHbEf3n0iRdYtnh2qXXu2LmHU6fOMPbNlyksLOS5ka/y0MBn2LBqUan9srNzCIqoT2pqGjWjIxn+zGAGD3rwcnePiIiIXKYxo4dz6lQK+/cfYsb0z4C/M8WGLXTo2Is77+zA7G8mkJmVxSuvvEvPnv1Zv/4ni22MHv0Od911O7NmWmaKzz6fQvv2rZg+7VM2b97Oa6+/T2FhIb+vWMtLLw3DxcWZ5577L48Nfp6lS+aUWufOnXsYP+4L3nnnZQoLjTz//Cs88sjTrFu3uNR+UDSJbl5eHlu37uKjj7/iiSf64+zsfEX76bJDWlRkNYICA4j/67jFZcFBj48gpkkDfpjztXnKjpvr1aZ+TEd+XrKcrl06mts2aXQzEz97z2rbYaFVmPb1RwB0vr0dP/60jI8+m8SebSuoU7to2O7xxBM88/wYzp5Nw9e35LnbzqadY/vGZQQFBZjfD35yJAkJiYSHhxbbp0ZUNd59azSNG9UjJyeXb+YsYMhTL5KWls4Lw4de7i4SERGRyxAVVY2goAD++ivB4nLhY48NJ6ZpQ76fN9kiUzRs1J6ff15O164XMkXjxvWLvccrLLQKU6d8DEDnzu1ZtHgpH3/yNX/sWkmdOjUBSDx+gmeHjb50pjibxtYtywgKCjS/f/zxF0rNFFAU0FzdIszvH3mkD+PHvXE5u8bCNT2XISsrm3UbYrnv3rsoLCykoKCAgoICatWMIqRKZWK37rRo37Vzx2K306nDbRbva9aIJDSksjmgAdSMjgQg4XhSqTU1alDPHNAA6v69jdL6PdSvFy8MH0rH9rfR7c5OzJr6Kb3v6cZb735Efn5+qZ8nIiIi1y4rK4t162Pp3fsfmaJWFCEhldmyZYdF+653Fp8pOna0zBTR0VGEhlYxB7SiZX9nioRLZIqG9cwBDaDu39u4VD8nJyc2bfyFVSsXMH7c6yxe/CuDHn2u1D7FuaaQlpp6lsLCQoa/+BrOXlUtXolJJziWkGjRPvii8HQxXx/LFOvi4myVbF1cik4R5uSUPIQVwNfX8jFNLi4uRf1KGfpanD69upOWdo5D/7j2LCIiItdfamoahYWFPD/iVVzdIixeiYnWmSIoOLDY7RSXH0rMBjk5pdbkc5VZBCAmphGtWzdn2LAhfPnle8yYMdcqaF7KNY3u9PX1wWAw8PKLz9Czexer9YGB/hbvy9MTDEp47ryIiIjcAOczxUujnqVHj+IyheWJnvKUKZo2aQjAocNxxMQ0uux+VxTSXFxcLNKjp6cHLZs35c99h3jrtYZXsim7N2feInx9fagRVc3WpYiIiPzruLg4W2eKFjH8ue8gb745yoaVXX9r120Crnz+1SsKaXVq1WDytG+ZPWcB0TWqExjoz/vvjKHDnX3p+9BQ7r+vB35+PiQcT+LX5asZ+Ehf2rW59YoKsoWYVnfS/8H7qFUziuzsHGZ9+wM/LPyZD99//YpHYoiIiMil1akdzZQp3zL72/l/Z4oA3n33v3S6/T7u7/c4ffv2wM/Xl4Tjifz222oG9L+fdu3sP1O0a9eTnj3vpHbtaBwcHNi0aRvjxn9B587tueWWxle0rSsKaY8O6MfmLTt4ZsQYUlJS6f/QfUz96kPWLp/Pq299wMDHnycvL4/wsBA6tm9NjchqV1SMrURVr8r4jydy4uQpDAaoX68OMyd/woP97rV1aSIiIv9KgwY9wObYHQwbNpqUlFQeeaQPUyZ/xOpVC3jt9Q949NHhRZkiPJQOHVpTo0Y1W5d8WZo2bcjXk2YRH5+As7Mz1avfxGuvvnBVT1YwmP5x89WhQ4eYNWsWw5/oh7d3petV87/O6vXbiN2+jxEjXrB1KSIiInZp5cqVbNu2heeGPWTrUuza7G9/wcnJg759+1osv6bRnSIiIiJyY5QY0jS2sXQa/CkiIiLXRQmhwiqknZ87JDc378YWVM7l5OaZ95WIiIhYc3Z2Jjc3T9NaXUJJmcIqpAUHB+Po6MiRuONWjeWCo/GJhIaU/EgIERGRii40NJS8vHwSE5NtXYrdysnJJTHxFKGh1pnCKqS5ubkRFRXFxi27STuXUSYFljc7dx/kZHIKdevVs3UpIiIidqtq1ap4enqyYkUsBQUFti7H7phMJlas2IzRaKROnTpW661GdwKcPXuWadOmkpubS91a1QkPDcbZyRHK0ey+15vJZCIrK4eDh//i0NEEGjVqxN13312uZjwWEREpa0ePHuWbb77By8uDunWjCAz0xcnR0dZl2ZTRaCItLZ0/9x0lKekU3bp1IyYmxqpdsSENIC0tjU2bNrF3717S0tJueMHlgcFg4KaICOrdfDMxMTEKaCIiIpfh+PHjxMbGsn///ks+L7OicHJyokaNGjRu3JiaNWsW26bEkHaxgoICCgsLr3uBV6Jv377MmTPHpjW4uLgomImIiFwlk8lEQUEBRqOxzD6zuPxg60xhMBhwdna+ZKa4rCcOODk54eR0Tc9iv2ZGoxFXV1eb1iAiIiJX73w4KUvF5Yfykik0ma2IiIiIHVJIExEREbFDCmkiIiIidkghTURERMQO2XY0wEXi4+PJz8+3WBYcHIy3t/cl+5pMJuLj4zVRnoiUC66urkRERFxWW5PJRFxcnM1H2IvYSmBgIL6+vpfVNisri8TERKD4DJGSklK+pgAx2QmKnulu8fLxrmRavny5yWQymbp3715i3zlz5hTbXy+99NLLXl+PPfaYKTc395I/G2fMmGHzWvXSy5Yvj0qVTD/++ONlZYkOnTqa+/kF+JtOnjxpzg/Z2dnmdbfddttlbc/W7OZMGsA3I8Dl74pMJvjslyy6dOnM5MlTSu2Xmppa9DSE3m+VQZUiItfodDyTpkxm3/4DLJj/AwEBASU2TU1NxdHJmdvesO08kSK2YeLQokn06NGDsWPHMnLkyFLnFks5cwYaRUKHxpyb8DOvv/66eV1eXp7532fOnLmhVV8vdhXSWtYC14umT7kl2sh/pht5+OGHqVWrFiaTqZT/OAaIqF8mdYqIXJOI+piCI9mw+H80bXYLv/y0uNjn9p1ncHQkuEGrMixQxH4E3Xwru2e9y3/+8x/27N3LxAkTSp/jzLcS3FyNwq4xTJg4kTa33WbV5OzZszeu4OvIrgcOODvBuIEw8h7Yv38/Q4YM0X1nIvLvEFaXwvvfJyHTSLNbmrNkyRJbVyRilwwODtR/+CWaj/iMb2Z/S/sOHUlOTr50x05NMPhWYu+ff1osDg5yISP97I0p9jqz65AGRVcxh98NHz4KU6dM4u7u3UhPT7d1WSIi186nCoV9xpJVuTZdu3Xjo48+wnTpJ/WJVEhV2/em7f/ms/PPgzRtdgu7du0qvYOzEwX33srJEydYs2aNeXHjhl6kncssF2fT7D6kndenNcwcbmLNquW0a9uapKQkW5ckInLtXDwwdX8ZU5MePPfccwwZMsTi3hmp2ArzytFIxDIQUKspHcYvIcvZi5a3tmLRokWld7ilNoaqlXluxPPmP4CaNCwa8bl169YbXe41KzchDaBNPVjwUiFJ8Xtp2aIZe/futXVJUk4NaVGVuNEdyfxfV34b2oJbbvLFNK47/ZuFm9v0bxbOzhFtyR7blYRXOvHWnbVxdDBYrDeN607jMB9+fqw5Ge/cyYFR7Xm4abjV53WtE8zGZ1uTNbYrya/fwee96uPh4lgm31XKAQdHaDMQOg9j0pSpdOx0OykpKbauqtxLi9/H6lcfYEG/2nzfqxq/PH4r++Z9CsDm/3uWJU+2IWnLcpY82YZ599zEr8NuJ2XfFottxC3/jt9f7M6C+2sxv29NVoy6h5T926w+69yxA6x7eyAL7q/F972qsfTp9vy16gfzepPJxL4fPufnIS2Z1zOCnx5txv4FX1psY/es9/mhd3VS9m9j+YiuzLvnJg4umnQD9kz55hEURrt3f8S/QRt69OjBe++9V/IZaAcDpr5t2Ra7hR9+KPrvER3lgYeHI7GxsWVY9dWxq4EDl6NuBCwaXcDDH56g1a0tWLDwEila5B+616vMhPsa8NXGeObtSqJxmA/fPNjEos3wNpG8d1cd/m/1EUYsOkWdYC/evrM2jg7w0k/7LNrOfKAxX236i/GrD/N4i6pMvb8RWxLO8ufJDAB6NQhhzsNNmRL7F68u3U+Itxtju9XBz92ZfjOtf9hLBVavIybfEIsBBXL11r75CG4+gTR7djzOnt6kJx4lO+XCVZic1GS2ffEf6j0wEudKPuyb+wmrX7mfOyduwM03CIDM5GNU7XAflapUw1iQz1+rfmDFqJ50/nQFXmFRAKQfP8LyF7rhERhK48ffxs0viLT4fWSeOm7+rO0TR3N02TfU6TOMgFpNOP3nFv6Y+haOLu7U6Nrf3M5YkM+mD56kZs/Hqd9/NC6VfMtmZ5UzTm6etHxpktWAgmLVjsDQuAZjXn0FAAcHA43qexEbu7kMK7465S6kAYT6w/xRhTz2WSZ33N6J+/s9YOuSpBwZ0yma5QdPMWRu0f0My/afws3Jgdc61wKgkqsjr3euxXsrDjP6l6JA9tuB0xQYjXzQvR7vrzjMmawLEy9/uu4oX6yPB2BjfCrd6lbm3vohvH3yIAAfdK/LnB3HGfzdhfsnTqbnsvjRW3jz1wPs/TvMiQAXBhT8+A4xtzTngfv7Fs3sJFckNy2FzBPxNB78JqHNOwMQ3KC1RZu89FRajvqKyg2LRv8F1WvJ4oGNObBwIg36jwagXr8R5vYmo5HKjdty5uAOjv72rbnNnm/ex8HJmQ7vL8bZwwuAyo3amvtlJMVxaPFkmj71HlFdHjGvL8jOZO/sD4jq8jAGh6ILW8aCfOr3f5mI1nffiN3yr3J+QIF3RE2++Xg4+/bv//tWAeurFKberUkcM838PqaxNz8s3lCG1V6dcnW582LeHjBzuJG7mxUwffp0MBltXZKUAw4GaBzmw497TlosX7jnhPnft1bzx8vNibm7EnF0MJhfvx9KwcPFkZurWM5gvWz/KfO/M3ILOXY2m3AfNwBqBnlSzd+D73YmWWxr1ZEUTEBMhO8N+65Sjp0fUBBYk6+//hqjUU8buFIu3v54BEewa9rbxC2fQ9bpRKs2zp7e5oAG4FLJh+AGrTmz/8K9SueOHWDdWwNY+FA95t4dwrweYaQnHCLj+BFzm+Sdawhv1d0c0P7p5I7VAITfehfGwgLzq3LD28hJTSbr9HGL9iExHa/pu1c0Vdv1ou3/5rNj7wH+/MdITrOQAGhdz/y2SSNvEhJOFN/WjpTLM2nnuTjBM3fBT1sgNx8U0+RSgiq54uzowKkMyxuzk9MvvA/0dAFg+/NtKU6Er5vF+7M5ltPC5BUYcXN2tNjWgoHNStiW+xVULxWKqwfc+gDEb7d1JeWSwWCgzRvfsnvGWLZ9MYqCnCz8ohrQaPAbBN3cEgBXb+tJhF19Akn/O4DlZ2Ww6r99cfUJoNGjr+MRHIGjiytbPn6ewvwLN/Tnpqfi7l+5xFpyz6WAycTCB4qfCy/rVCKewUWPCXN0dcfJzfOqv3dF5V+zCdU69uXAjyVc8gSoHQFrdgNQOdiljCq7NuU6pG06AIM+ccTX35+kE6dtXY6UA6cycskvNBJUyfJ/0GCvC+/PX8q8Z0osx85mW23j6Jmsy/6889t66oc/2BSfarU+8ZxGbkkJDqzDYemHVAkLJTlZP9+uhnd4DW596WuMBfmc/jOWP6a/w9o3HuauaTuBv8PTP+SmncbdPxiAlH1byD6dyG2vzMQ38sJZmPysc7gTYn7v6uVH9pmTVts6z8XLDwwGOry3CAcnZ6v1XmE1zP8ubTZ9KV5hfi5bP3mBuN+/I6hKZU4V18hoxOGnWPPJnF2703Fxsf5vYW/K7eXORbFw/wcONGx6Ky+MHFU0oZrIJRhNsP14Gj3qVbFY3vPmC+/Xx50hM7eAcF83tiakWb0uvh/tUvYlZ3DsbDaR/h7FbivpXO51+27yL2EywYZvYfG79L63JyOGDwf9eLsmDk7OBNe/lTq9nyE/K52cM0WXufIzz3Fy54X5s/Iyz5G8ay3+NYsGEhXmFf2R5uB84Zf56T9jyTx5zGL7wY3akLBuEflZxd9fev6Sat65M/hHN7J6OXtUun5ftoLJOXuK1aN7kbhuIbNmzSI0NLT4huv2YEy88MfOtp3pNGxg/08pKnchzWSCCUvh8c+hV+8+LFn6K56eOjUsl++t3w7SITqQifc14PaaQbzYPso8bYbRCOdyCnhl6X7eu6suY7vVoXOtIG6vGcTjLavy82PNcXe+sqkznl+4h2dvq84XvepzV93KtK8RwIBmEcx9pCnRgTp25SL5uRh+GQcbvuGNN97g29mzcXa2/7/27dHZo3tYNeY+jiydSfKutRzf8DN75/wfnpUj8KxSDSg6w7Xl4+HELf+O45uWsObVfgDU7DEEKJqTy8ndk21fjOLEtpUc/XU2G98bintAiMVn1ev3AsaCfH5/sTvxK+ZxcucaDi6aZJ7uwyssihrdBrJp/NPsnfN/nNyxiqQtyzmwcCJr3+qPXJ2zR/ew4vkukHKM1atW8cADJQwizM3HacEG7u3Vy7xo+84smt3SoowqvXrl6nJnoRFemw2TfoNRo0bx9ttv4+BQ7nKm2NiiPScZOm8XL3eM5qGm4WyKT+XJ7/9gyZAWpOUUnSUbv+oIx9NyeL5tJM+0rk5+oZHDKVks3nuSvMIru/tx3q4kzn69idF/fx5A3Jksluw/xckMnUmTv2Wk4LjofzilHmPm3Ln07t3b1hWVa25+wbj5BfPn3I/JTjmBs6cXQXVb0HzE5zg4OprbNBj4X3ZNfoOMpDi8q9aizRtzcPMLNq9vOeprdk56jXVv9adSaCRNn3rPHL7O8wqLpMP7i/lj2tts+2IUxsICvMKiqN37GXObxo+/g1dYDY4smc7e2eNxcvPAKyyK8Ns0ivNqHN/4C5vHPUXtmtH8tOhHIiIiSm68dAtk5vDaq6/yw/ffcy69gIOHz/FSs+LvFbYn5SakZeXCM18ZWLodvvjic4YOHWrrkqQcm7Ahngkb4s3vH21+EwA7E8+Zl83ZkcicHdYjws6bFpvAtNgEq+X1P1hltey3A6f57YDuK5ISnDyM06K3CfB04ae1a2jatKmtKyr33HyDaD7is0u2C212O6HNbi9xfUjTDoQ07WC5rJjRlz431aL1f6eXuB2DwUB090eJ7v5oiW1ufnAkNz848pI1V2Qmk4l933/K7mlv07PnPcyYMb30q2lpmTgs2cKzTz9D1apVAdj5RzomEzRTSLs+Us7BgE8c+fO4Mz/+OI9u3brZuiQpx/zcnXn1jpr8fug06bkFNIvwZXSnaBbsPkF8qvVAAZEb6u8BAg0a1GfRjwtLvqdGpIK7eIDAyy+/zJtvvnnpq2kL1lPJzZ0xY8aYF23beQ5PT3dq1659gyu+dnYf0o6cgIc+dCTL6MOq1UuJiYmxdUlSzuUbjUQFetKvcRh+Hs6cyshjxtYE/rO4hPl1RG4Ekwk2zoEN39C7b1+mTpmCu7umZBEpTk5qMhveGUjakT+YNWtWyfefXez4aQxr/uD1cePx8/Pj3LmiKyXbd6bTpk0rHB3t/9F8dh3SthyCAR87ElSlGr8v/ZXq1avbuiT5F8jILaT7JPt/HIj8i+XnYvj1E0z7VvPGG28wZswYTb1Qxm4Z/rGtS5DLdPbIHja89QhuhgJWr1pF8+bNL6ufw7w1hN9UlSeffNJqXbNm9j9oAOwspD32KZw/c2kywdo/HYhpdgsLf1yMv79/6Z1NRljw1o0vUkTkGjmkJeGcefqyBwgU5uWy9o2Hy6AyEftiMpk4vXv95Q0QOO+PozD+e4y74/jgu+9wcbGeuLY83I8GdhTS7r33XvLzLeeferJjNG+//TZubm4l9CrStm1b7rnnHgoKCkptJyJiD1xdb2LUqFGXNUCgffv29OjRg8JCPRpKKqbqLR/lnXfeuazpth5/bDC//PILAPXuut/ijyAvLy+GDh1Kamoqd9xxxw2r93oymEwmPbpXRERExM5okjERERERO6SQJiIiImKHFNJERERE7JDdhLSCggJWrlypm/9LoP1zadpHpdP+uTTto9Jp/1ya9lHptH+ujN2EtMLCQlatWqURTCXQ/rk07aPSaf9cmvZR6bR/Lk37qHTaP1fGbkKaiIiIiFygkCYiIiJihxTSREREROyQ3YQ0R0dH2rZtWy4eeGoL2j+Xpn1UOu2fS9M+Kp32z6VpH5VO++fK6IkDIiIiInbIbs6kiYiIiMgFCmkiIiIidkghTURERMQOKaSJiIiI2CEnWxcAEBsby/r160lPTyc4OJjOnTtTtWpVW5d1w8XHx7N+/XoSExPJyMigb9++1K5d27x+wYIF7Ny506JPWFgYjz32mPl9QUEBy5YtY/fu3RQUFFC9enW6deuGt7d3mX2PGyU2NpYtW7Zw9uxZAIKDg2nTpg3R0dEAmEwmVq1axdatW8nJySEsLIyuXbsSHBxs3sa/ef8UZ82aNfz+++80b96cLl26ADqOzp07x2+//cahQ4fIz88nICCAu+++m9DQUEDH0YcffkhaWprV8piYGLp161bhjx+A3NxcVqxYwb59+8jMzKRKlSp06dKFsLAwoOIcQ5f6nVWcuLg4li1bRnJyMl5eXrRq1YqYmBjz+h07drBw4UKrfqNHj8bJ6UJEqag5weajO3fv3s38+fPp1q0bERERbN26lW3btvHUU0/h4+Njy9JuuIMHD3Ls2DFCQkL47rvvig1pmZmZ9OjRw7zM0dERd3d38/vFixdz4MABevbsibu7O8uWLSM7O5shQ4bg4FC+T5Tu378fBwcH/P39gaL/mdevX8/jjz9OcHAwa9euZc2aNfTs2ZOAgABWr15NfHw8Tz/9NK6ursC/e//80/Hjx5k3bx6urq5Uq1bNIqRV1OMoOzubCRMmUL16dWJiYvD09OTMmTP4+vqaj6uKfhxlZmZy8a+B5ORkZsyYQf/+/alWrVqFPn7OmzdvHsnJyXTr1g0vLy927drFxo0befLJJ/H29q4wx9Clfmf9U2pqKl988QVNmjShadOmHDt2jJ9++olevXpRt25doOjn+pIlS3j66act+laqVMn874qcE2x+ZGzcuJHGjRvTpEkTgoKC6NKlCz4+PsTGxtq6tBsuOjqaDh06UKdOnRLbODo6UqlSJfPr4h+MOTk5bN++nTvuuIPIyEhCQkK49957SU5O5siRI2XxFW6oWrVqER0dTUBAAAEBAXTs2BEXFxcSEhIwmUxs2rSJ2267jTp16hAcHEzPnj3Jz8/njz/+AP79++dieXl5/PDDD3Tv3h03Nzer9RX1OFq3bh0+Pj706NGDsLAwfH19iYyMNAc0HUfg6elpcWwcOHAAPz8/i7MUFfX4AcjPz2fv3r106tSJqlWr4u/vT7t27fD19WXLli0V6hi6nN9ZF9uyZQs+Pj506dKFoKAgmjRpQuPGjdmwYYNV24uPr4sDGlTsnGDTkFZYWEhiYiJRUVEWyyMjI0lISLBRVfYlLi6O999/n08++YQff/yRzMxM87qkpCSMRqPF/vPy8iI4OJhjx47Zotwbxmg0snv3bvLz84mIiODs2bNkZGRYfHcnJyeqVatmPnYq0v75+eefiY6OJjIystj1FfU42r9/PyEhIcydO5f333+fCRMmsHXrVvN6HUeWCgsL2bVrF40bN8ZgMJiXV9TjB4p+9phMJotLbwDOzs789ddfOoZKkZCQYPUzKSoqisTERIsHrOfl5fHhhx8yfvx4vvnmG5KSkszrKnpOsOk9aVlZWZhMJqvUXKlSJQ4fPmyjquxHjRo1qFu3Lr6+vqSmprJixQqmTZvGkCFDcHJyIiMjw+qyAxT9ZZyRkWGjqq+vkydPMmnSJAoKCnBxcaFv374EBQWZf7D989jx9PQ0319TEfYPFF0KSEpKYvDgwcWur8jHUWpqKlu2bKFly5a0bt2a48ePs2TJEpycnGjYsKH5++k4KrJv3z5ycnJo1KiReVlFPn4AXF1dCQ8PZ/Xq1QQFBeHp6cnu3btJSEggICBAx1ApMjIyiv39bjQaycrKwsvLi8DAQHr27ElwcDC5ubls2rSJyZMnM3ToUAICAip8TrCLgQP/pIcgFLn55pvN/w4ODiY0NJQPP/yQgwcPXvJ088V/BZdngYGBDB06lJycHPbu3cuCBQsYMGBAie0v99j5t+yftLQ0lixZwkMPPWT1l/55Ffk4MplMhIaG0rFjRwBCQkI4deoUW7ZsoWHDhqX2uxzlff/80/bt24mOjsbLy8u8rCIfP+fdc889/Pjjj4wfPx6DwUBISAj169e3OOPzTxX1GLqU8/vl/PcODw8nPDzcvP6mm25iwoQJbN68mTvvvPOS2/m3s2lI8/DwwGAwWP0lkZmZaZWapej0uK+vLykpKUDRXxKFhYVkZ2db/IWWmZlpcdCXZ46Ojub7h0JDQ0lMTGTjxo20bt0aKPpL7eJfKFlZWeZjpyLsn6SkJDIzM5k4caJ5mclkIj4+ns2bNzNmzBirm5Ir0nHk5eVFUFCQxbLAwED+/PNP4MLZj4p+HEHRpd8jR47Qp0+fUttVpOPnPH9/fwYMGEBeXh65ubl4eXkxb948/Pz8dAyVolKlSsX+fndwcLA6q3iewWAgNDSUM2fOAMoJNr0nzdHRkdDQUKsbJ48cOfKvPnCvVlZWFmlpaeYfBCEhITg4OFjsv/T0dJKTk4mIiLBVmTdcYWEhvr6+VKpUyeK7FxYWEhcXZz52KsL+qV69Ok888QRDhw41v0JDQ2nQoAFDhw4tdtRYRTqOIiIizGHivJSUFPOIMB1HF+zYsQNPT09q1qxZaruKdPz8k4uLC15eXmRnZ3Po0CFq1aqlY6gU4eHhVr/fDx8+TGhoaIkPWDeZTJw8edIcwCp6TrD55c4WLVowf/58QkNDCQ8PZ+vWraSlpVnMo/JvlZeXZ/5rAYrunzlx4gTu7u64u7uzcuVK6tSpg5eXF2fPnmX58uV4eHiYhzy7ubnRuHFjli1bZu7z66+/EhwcXOIN5OXJ8uXLqVGjBj4+PuTm5rJ7927i4uJ48MEHMRgMNG/enDVr1uDv709AQABr1qzB2dmZ+vXrA//+/QNF98tcPBcTFN3Q7O7uTnBwMHl5eRX6OGrRogWTJ09mzZo11KtXj+PHj7Nt2zbuuusuAB1HfzOZTOzYsYOGDRtaBPuKfvycd+jQIQACAgI4c+YMv/76K4GBgTRq1KhCHUOl/c4qbiqMmJgYYmNjWbp0KU2aNCEhIYHt27fTq1cvc5uVK1cSHh5OQECA+Z60EydO0LVrV3ObipwTbD5PGhRNUrdu3ToyMjIq1CR1cXFxTJs2zWp5w4YN6datG3PmzCEpKYmcnBy8vLyoVq0a7du3t/if4eIJEvPz84mMjKRr167/irljFi5cyNGjR8nIyMDV1ZXKlSvTqlUr8yifiyeQzM7OJjw8vNQJJP9t+6ckU6dONU+2mZ+fX+GPowMHDrB8+XJSUlLw8/OjRYsWNG3a1Lxex1HR2Y2ZM2fy9NNPExAQYF6u46fInj17WL58OefOncPd3Z06derQoUMH83Q3FeUYKu13Vs+ePVm5ciU7duzgueees+izdOlSTp06VexktkuWLGHfvn3mn/MhISG0bdvW6gxjRc0JdhHSREREpHxbsGABAD179rRpHf8mNp/MVkRERMq/+Ph42rdvb+sy/lV0Jk1ERETEDulMmoiIiIgdUkgTERERsUMKaSIiIiJ2SCFNRERExA4ppImIiIjYIYU0ERERETukkCYiIiJihxTSREREROyQQpqIiIiIHVJIExEREbFD/w81oiKshaoahgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 700x200 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = targets.plot(ax_width=7,\n",
    "                   plots_indexing='biopython',  # numbering starts at 0\n",
    "                   ax_height=2,  # height of each plot\n",
    "                   hspace=1.2,  # vertical space between plots\n",
    "                   )\n",
    "\n",
    "plotfile = os.path.join(config['process_ccs_dir'], 'amplicons_'+background+'.pdf')\n",
    "print(f\"Saving plot to {plotfile}\")\n",
    "fig.savefig(plotfile, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write out the specs used to parse the features (these are the same specs provided as `feature_parse_specs` when initializing `targets`, but with defaults filled in):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MERS:\n",
      "  query_clip5: 5\n",
      "  query_clip3: 5\n",
      "  termini5:\n",
      "    filter:\n",
      "      clip5: 4\n",
      "      mutation_nt_count: 1\n",
      "      mutation_op_count: null\n",
      "      clip3: 0\n",
      "    return: []\n",
      "  gene:\n",
      "    filter:\n",
      "      mutation_nt_count: 90\n",
      "      mutation_op_count: null\n",
      "      clip5: 0\n",
      "      clip3: 0\n",
      "    return:\n",
      "    - mutations\n",
      "    - accuracy\n",
      "  spacer:\n",
      "    filter:\n",
      "      mutation_nt_count: 1\n",
      "      mutation_op_count: null\n",
      "      clip5: 0\n",
      "      clip3: 0\n",
      "    return: []\n",
      "  barcode:\n",
      "    filter:\n",
      "      mutation_nt_count: 0\n",
      "      mutation_op_count: null\n",
      "      clip5: 0\n",
      "      clip3: 0\n",
      "    return:\n",
      "    - sequence\n",
      "    - accuracy\n",
      "  termini3:\n",
      "    filter:\n",
      "      clip3: 4\n",
      "      mutation_nt_count: 1\n",
      "      mutation_op_count: null\n",
      "      clip5: 0\n",
      "    return: []\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(targets.feature_parse_specs('yaml'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align CCSs to amplicons\n",
    "We now align the CCSs to the amplicon and parse features from the resulting alignments using the specs above.\n",
    "\n",
    "First, we initialize an `alignparse.minimap2.Mapper` to align the reads to SAM files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using `minimap2` 2.26-r1175 with these options:\n",
      "-A2 -B4 -O12 -E2 --end-bonus=13 --secondary=no --cs\n"
     ]
    }
   ],
   "source": [
    "mapper = alignparse.minimap2.Mapper(alignparse.minimap2.OPTIONS_CODON_DMS)\n",
    "\n",
    "print(f\"Using `minimap2` {mapper.version} with these options:\\n\" +\n",
    "      ' '.join(mapper.options))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we use `Targets.align_and_parse` to create the alignments and parse them:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "readstats, aligned, filtered = targets.align_and_parse(\n",
    "        df=pacbio_runs,\n",
    "        mapper=mapper,\n",
    "        outdir=config['process_ccs_dir'],\n",
    "        name_col='run',\n",
    "        group_cols=['name', 'library'],\n",
    "        queryfile_col='fastq',\n",
    "        overwrite=True,\n",
    "        ncpus=config['max_cpus'],\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, examine the read stats from the alignment / parsing, both extracting alignment target name and getting stats aggregated by target:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "readstats = (\n",
    "    readstats\n",
    "    .assign(category_all_targets=lambda x: x['category'].str.split().str[0],\n",
    "            target=lambda x: x['category'].str.split(None, 1).str[1],\n",
    "            valid=lambda x: x['category_all_targets'] == 'aligned')\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now plot the read stats by run (combining all targets and libraries within a run):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncol = 2\n",
    "p = (\n",
    "    ggplot(readstats\n",
    "           .groupby(['name', 'category_all_targets', 'valid'])\n",
    "           .aggregate({'count': 'sum'})\n",
    "           .reset_index(),\n",
    "           aes('category_all_targets', 'count', fill='valid')) +\n",
    "    geom_bar(stat='identity') +\n",
    "    facet_wrap('~ name', ncol=ncol) +\n",
    "    theme(axis_text_x=element_text(angle=90),\n",
    "          figure_size=(1.85 * min(ncol, len(pacbio_runs)),\n",
    "                       2 * math.ceil(len(pacbio_runs) / ncol)),\n",
    "          panel_grid_major_x=element_blank(),\n",
    "          legend_position='none',\n",
    "          ) +\n",
    "    scale_fill_manual(values=CBPALETTE)\n",
    "    )\n",
    "_ = p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And the read stats by library (combining all targets and runs within a library):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = (\n",
    "    ggplot(readstats\n",
    "           .groupby(['library', 'category_all_targets', 'valid'])\n",
    "           .aggregate({'count': 'sum'})\n",
    "           .reset_index(), \n",
    "           aes('category_all_targets', 'count', fill='valid')) +\n",
    "    geom_bar(stat='identity') +\n",
    "    facet_wrap('~ library', nrow=1) +\n",
    "    theme(axis_text_x=element_text(angle=90),\n",
    "          figure_size=(1.5 * pacbio_runs['library'].nunique(), 2),\n",
    "          panel_grid_major_x=element_blank(),\n",
    "          legend_position='none',\n",
    "          ) +\n",
    "    scale_fill_manual(values=CBPALETTE)\n",
    "    )\n",
    "_ = p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And the number of reads by target (combining all libraries and runs for a target):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = (\n",
    "    ggplot(readstats\n",
    "           .groupby(['target'])\n",
    "           .aggregate({'count': 'sum'})\n",
    "           .reset_index(), \n",
    "           aes('target', 'count')) +\n",
    "    geom_point(stat='identity', size=3) +\n",
    "    theme(axis_text_x=element_text(angle=90),\n",
    "          figure_size=(0.3 * readstats['target'].nunique(), 2),\n",
    "          panel_grid_major_x=element_blank(),\n",
    "          ) +\n",
    "    scale_y_log10(name='number of reads')\n",
    "    )\n",
    "_ = p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And read stats by target (combining all libraries and runs for a target):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = (\n",
    "    ggplot(readstats\n",
    "           .groupby(['target', 'valid'])\n",
    "           .aggregate({'count': 'sum'})\n",
    "           .reset_index()\n",
    "           .assign(total=lambda x: x.groupby('target')['count'].transform('sum'),\n",
    "                   frac=lambda x: x['count'] / x['total'],\n",
    "                   ), \n",
    "           aes('target', 'frac', fill='valid')) +\n",
    "    geom_bar(stat='identity') +\n",
    "    theme(axis_text_x=element_text(angle=90),\n",
    "          figure_size=(0.5 * readstats['target'].nunique(), 2),\n",
    "          panel_grid_major_x=element_blank(),\n",
    "          ) +\n",
    "    scale_fill_manual(values=CBPALETTE)\n",
    "    )\n",
    "_ = p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's see **why** we filtered the reads.\n",
    "First, we do some transformations on the `filtered` dict returned by `Targets.align_and_parse`.\n",
    "Then we count up the number of CCSs filtered for each reason, and group together \"unusual\" reasons that represent less than some fraction of all filtering.\n",
    "For now, we group together all targets to the stats represent all targets combined:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "other_cutoff = 0.02  # group as \"other\" reasons with <= this frac\n",
    "\n",
    "filtered_df = (\n",
    "    pd.concat(df.assign(target=target) for target, df in filtered.items())\n",
    "    .groupby(['library', 'name', 'run', 'filter_reason'])\n",
    "    .size()\n",
    "    .rename('count')\n",
    "    .reset_index()\n",
    "    .assign(tot_reason_frac=lambda x: (x.groupby('filter_reason')['count']\n",
    "                                       .transform('sum')) / x['count'].sum(),\n",
    "            filter_reason=lambda x: numpy.where(x['tot_reason_frac'] > other_cutoff,\n",
    "                                                x['filter_reason'],\n",
    "                                                'other')\n",
    "            )\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now plot the filtering reason for all runs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncol = 7\n",
    "nreasons = filtered_df['filter_reason'].nunique()\n",
    "\n",
    "p = (\n",
    "    ggplot(filtered_df, aes('filter_reason', 'count')) +\n",
    "    geom_bar(stat='identity') +\n",
    "    facet_wrap('~ name', ncol=ncol) +\n",
    "    theme(axis_text_x=element_text(angle=90),\n",
    "          figure_size=(0.25 * nreasons * min(ncol, len(pacbio_runs)),\n",
    "                       2 * math.ceil(len(pacbio_runs) / ncol)),\n",
    "          panel_grid_major_x=element_blank(),\n",
    "          )\n",
    "    )\n",
    "_ = p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we take the successfully parsed alignments and read them into a data frame, keeping track of the target that each CCS aligns to.\n",
    "We also drop the pieces of information we won't use going forward, and rename a few columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First few lines of information on the parsed alignments:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>library</th>\n",
       "      <th>query_name</th>\n",
       "      <th>gene_mutations</th>\n",
       "      <th>gene_accuracy</th>\n",
       "      <th>barcode</th>\n",
       "      <th>barcode_accuracy</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/51/ccs</td>\n",
       "      <td>C31A A33T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>TCAACTTTACCAGGTG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/143/ccs</td>\n",
       "      <td>C571G T572A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>CTTATCCCTTAACAGT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/152/ccs</td>\n",
       "      <td>del134to136</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>CCATATCCCGCTCTTG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/184/ccs</td>\n",
       "      <td>C502A G504A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>AACGGAAAAAAACCTT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/242/ccs</td>\n",
       "      <td>A118T C120T</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>ATGTTCGTCCCAGATA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "aligned_df = (\n",
    "    pd.concat(df.assign(target=target) for target, df in aligned.items())\n",
    "    .drop(columns=['query_clip5', 'query_clip3', 'run','name'])\n",
    "    .rename(columns={'barcode_sequence': 'barcode'})\n",
    "    )\n",
    "\n",
    "print(f\"First few lines of information on the parsed alignments:\")\n",
    "display(HTML(aligned_df.head().to_html(index=False)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write valid CCSs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write the processed CCSs to a file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Barcodes and mutations for valid processed CCSs have been written to results/process_ccs/processed_ccs_MERS.csv.\n"
     ]
    }
   ],
   "source": [
    "aligned_df.to_csv(config['processed_ccs_file' + '_' + background], index=False)\n",
    "\n",
    "print(\"Barcodes and mutations for valid processed CCSs \"\n",
    "      f\"have been written to {config['processed_ccs_file' + '_' + background]}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we analyze these processed CCSs to build the variants."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build barcode variant table\n",
    "Builds consensus sequences for barcoded variants from the mutations called in the processed PacBio CCSs.\n",
    "Uses these consensus sequences to build a codon variant table."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make output directories if needed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(config['variants_dir'], exist_ok=True)\n",
    "os.makedirs(config['figs_dir'], exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the CSV file with the processed CCSs into a data frame, display first few lines:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 353832 CCSs from 2 libraries and 1 targets.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>library</th>\n",
       "      <th>query_name</th>\n",
       "      <th>gene_mutations</th>\n",
       "      <th>gene_accuracy</th>\n",
       "      <th>barcode</th>\n",
       "      <th>barcode_accuracy</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/51/ccs</td>\n",
       "      <td>C31A A33T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>TCAACTTTACCAGGTG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/143/ccs</td>\n",
       "      <td>C571G T572A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>CTTATCCCTTAACAGT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/152/ccs</td>\n",
       "      <td>del134to136</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>CCATATCCCGCTCTTG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/184/ccs</td>\n",
       "      <td>C502A G504A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>AACGGAAAAAAACCTT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/242/ccs</td>\n",
       "      <td>A118T C120T</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>ATGTTCGTCCCAGATA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "processed_ccs = pd.read_csv(config['processed_ccs_file' + '_' + background], na_filter=None)\n",
    "\n",
    "nlibs = processed_ccs['library'].nunique()  # number of unique libraries\n",
    "\n",
    "ntargets = processed_ccs['target'].nunique()  # number of unique targets\n",
    "\n",
    "print(f\"Read {len(processed_ccs)} CCSs from {nlibs} libraries and {ntargets} targets.\")\n",
    "\n",
    "display(HTML(processed_ccs.head().to_html(index=False)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Optional: discard reads that have a `fwd` or `rev` indicator suggestive of heteroduplexes (or we suspect in these runs, some form of pseudo-duplex)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#start = len(processed_ccs)\n",
    "#processed_ccs = processed_ccs.loc[~processed_ccs['query_name'].str.contains('fwd')]\n",
    "#processed_ccs = processed_ccs.loc[~processed_ccs['query_name'].str.contains('rev')]\n",
    "#end = len(processed_ccs)\n",
    "\n",
    "#print(f\"Went from {start} CCSs to {end} after discarded split/duplexed fwd and rev reads.\")\n",
    "\n",
    "#display(HTML(processed_ccs.head().to_html(index=False)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overall statistics on number of total CCSs and number of unique barcodes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>total_CCSs</th>\n",
       "      <th>unique_barcodes</th>\n",
       "      <th>avg_CCSs_per_barcode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>target</th>\n",
       "      <th>library</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">MERS</th>\n",
       "      <th>lib51</th>\n",
       "      <td>170624</td>\n",
       "      <td>59682</td>\n",
       "      <td>2.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lib52</th>\n",
       "      <td>183208</td>\n",
       "      <td>80134</td>\n",
       "      <td>2.29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(\n",
    "    processed_ccs\n",
    "    .groupby(['target', 'library'])\n",
    "    .aggregate(total_CCSs=('barcode', 'size'),\n",
    "               unique_barcodes=('barcode', 'nunique'))\n",
    "    .assign(avg_CCSs_per_barcode=lambda x: x['total_CCSs'] / x['unique_barcodes'])\n",
    "    .round(2)\n",
    "    .to_html()\n",
    "    ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter processed CCSs\n",
    "We have the PacBio `ccs` program's estimated \"accuracy\" for both the barcode and the gene sequence for each processed CCS.\n",
    "We will filter the CCSs to only keep ones of sufficiently high accuracy.\n",
    "\n",
    "First, we want to plot the accuracies.\n",
    "It is actually visually easier to look at the error rate, which is one minus the accuracy.\n",
    "Because we want to plot on a log scale (which can't show error rates of zero), we define a *error_rate_floor*, and set all error rates less than this to that value:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "error_rate_floor = 1e-7  # error rates < this set to this\n",
    "if error_rate_floor >= config['max_error_rate']:\n",
    "    raise ValueError('error_rate_floor must be < max_error_rate')\n",
    "\n",
    "processed_ccs = (\n",
    "    processed_ccs\n",
    "    .assign(barcode_error=lambda x: numpy.clip(1 - x['barcode_accuracy'],\n",
    "                                               error_rate_floor, None),\n",
    "            gene_error=lambda x: numpy.clip(1 - x['gene_accuracy'],\n",
    "                                            error_rate_floor, None)\n",
    "            )\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now plot the error rates, drawing a dashed vertical line at the threshold separating the CCSs we retain for consensus building versus those that we discard:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = (\n",
    " ggplot(processed_ccs\n",
    "        .melt(value_vars=['barcode_error', 'gene_error'],\n",
    "              var_name='feature_type', value_name='error rate'),\n",
    "        aes('error rate')) +\n",
    " geom_histogram(bins=25) +\n",
    " geom_vline(xintercept=config['max_error_rate'],\n",
    "            linetype='dashed',\n",
    "            color=CBPALETTE[1]) +\n",
    " facet_wrap('~ feature_type') +\n",
    " theme(figure_size=(4.5, 2)) +\n",
    " ylab('number of CCSs') +\n",
    " scale_x_log10()\n",
    " ).draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Flag the CCSs to retain, and indicate how many we are retaining and purging due to the accuracy filter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "processed_ccs = (\n",
    "    processed_ccs\n",
    "    .assign(retained=lambda x: ((x['gene_error'] < config['max_error_rate']) &\n",
    "                                (x['barcode_error'] < config['max_error_rate'])))\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here are number of retained CCSs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = (\n",
    " ggplot(processed_ccs.assign(xlabel=lambda x: x['target'] + ', ' + x['library'])\n",
    "                     .groupby(['xlabel', 'retained'])\n",
    "                     .size()\n",
    "                     .rename('count')\n",
    "                     .reset_index(),\n",
    "        aes('xlabel', 'count', color='retained', label='count')) +\n",
    " geom_point(size=3) +\n",
    " geom_text(va='bottom', size=7, ha='center',format_string='{:.3g}', nudge_y=0.2) +\n",
    " theme(figure_size=(0.5 * nlibs * ntargets, 3),\n",
    "       panel_grid_major_x=element_blank(),\n",
    "       axis_text_x=element_text(angle=90),\n",
    "       ) +\n",
    " scale_y_log10(name='number of CCSs') +\n",
    " xlab('') +\n",
    " scale_color_manual(values=CBPALETTE[1:])\n",
    " ).draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequences per barcode\n",
    "How many times is each barcode sequenced?\n",
    "This is useful to know for thinking about building the barcode consensus.\n",
    "\n",
    "Plot the distribution of the number of times each **barcode** is observed among the retained CCSs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_count = 8 # in plot, group all barcodes with >= this many counts\n",
    "\n",
    "p = (\n",
    " ggplot(\n",
    "    processed_ccs\n",
    "     .query('retained')\n",
    "     .groupby(['library', 'barcode'])\n",
    "     .size()\n",
    "     .rename('nseqs')\n",
    "     .reset_index()\n",
    "     .assign(nseqs=lambda x: numpy.clip(x['nseqs'], None, max_count)),\n",
    "    aes('nseqs')) +\n",
    " geom_bar() +\n",
    " facet_wrap('~ library', nrow=1) +\n",
    " theme(figure_size=(1.75 * nlibs, 2),\n",
    "       panel_grid_major_x=element_blank(),\n",
    "       ) +\n",
    " ylab('number of barcodes') +\n",
    " xlab('CCSs for barcode')\n",
    " )\n",
    "\n",
    "_ = p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Empirical accuracy of CCSs\n",
    "We want to directly estimate the accuracy of the gene-barcode link rather than relying on the PacBio `ccs` accuracy, which doesn't include inaccuracies due to things like strand exchange or the same barcode on different sequences.\n",
    "\n",
    "One way to do this is to examine instances when we have multiple sequences for the same barcode. \n",
    "We can calculate the empirical accuracy of the sequences by looking at all instances of multiple sequences of the same barcode and determining how often they are identical.\n",
    "This calculation is performed by `alignparse.consensus.empirical_accuracy` using the equations described in the docs for that function.\n",
    "\n",
    "We will do this four for sets of sequences:\n",
    "\n",
    " 1. All of the CCSs retained above.\n",
    " 2. CCSs retained by applying a PacBio `ccs` accuracy filter 10-fold more stringent than the one above.\n",
    "    The rationale is that if this improves the concordance (real accuracy) of the CCSs substantially then maybe we should make the accuracy filter more stringent.\n",
    " 3. Like (1) but excluding all CCSs with indels.\n",
    "    the rationale is that we only really care about substitutions, and will exclude sequences with indels anyway.\n",
    " 4. Like (2) but excluding all CCSs with indels.\n",
    " \n",
    "First, we annotate the sequences with the number of indels and whether they have an indel to enable categorization into the aforementioned sets. We want to convert the way we call *only* in-frame deletions to be substitution style (that is, `del511to513` becomes `C511- C512- G513-`), if the deletion is a single-codon in-frame deletion. We do this using alignparse util InFrameDeletionsToSubs. Below, when we parse indels, these in-frame deletions will be treated as \"substitutions\" and not add to indel counts, so we can then parse on indel count which will be count of out-of-frame indels.:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read gene of 645 nts for MERS from data/PacBio_amplicon_MERS.gb\n"
     ]
    }
   ],
   "source": [
    "targets = alignparse.targets.Targets(seqsfile=config['amplicons_' + background],\n",
    "                                     feature_parse_specs=config['feature_parse_specs_' + background])\n",
    "geneseq = targets.get_target(background).get_feature('gene').seq\n",
    "\n",
    "print(f\"Read gene of {len(geneseq)} nts for {background} from {config['amplicons_' + background]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>query_name</th>\n",
       "      <th>gene_mutations</th>\n",
       "      <th>gene_accuracy</th>\n",
       "      <th>barcode</th>\n",
       "      <th>barcode_accuracy</th>\n",
       "      <th>target</th>\n",
       "      <th>barcode_error</th>\n",
       "      <th>gene_error</th>\n",
       "      <th>retained</th>\n",
       "      <th>gene_mutations_with_del</th>\n",
       "      <th>n_indels</th>\n",
       "      <th>has_indel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/51/ccs</td>\n",
       "      <td>C31A A33T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>TCAACTTTACCAGGTG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>C31A A33T</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/143/ccs</td>\n",
       "      <td>C571G T572A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>CTTATCCCTTAACAGT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>C571G T572A</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/152/ccs</td>\n",
       "      <td>del134to136</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>CCATATCCCGCTCTTG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>G136- A137- T138-</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/184/ccs</td>\n",
       "      <td>C502A G504A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>AACGGAAAAAAACCTT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>C502A G504A</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/242/ccs</td>\n",
       "      <td>A118T C120T</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>ATGTTCGTCCCAGATA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>5.087173e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>A118T C120T</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/277/ccs</td>\n",
       "      <td>T82A T83C C84T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>AAGCAACTAAAAAACG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>T82A T83C C84T</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/326/ccs</td>\n",
       "      <td>A109G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>CCAGAGTTAAGACATT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>A109G</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/335/ccs</td>\n",
       "      <td>T250C C251A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>ATTTACTAATCAATAA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>T250C C251A</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/352/ccs</td>\n",
       "      <td>A190T G191T T192C</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>CAAAGGATTCATAGCC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>A190T G191T T192C</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/420/ccs</td>\n",
       "      <td>G235T A236T C237T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>ATGGAGGGCATTAGGC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>G235T A236T C237T</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/465/ccs</td>\n",
       "      <td>A488C</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>TCGCATGGGACTTAAG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>A488C</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>lib51</td>\n",
       "      <td>m64272e_230126_100200/512/ccs</td>\n",
       "      <td>G43C G44A C45A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>ATAGATAAGAAATAAG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MERS</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>G43C G44A C45A</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   library                     query_name     gene_mutations  gene_accuracy  \\\n",
       "0    lib51   m64272e_230126_100200/51/ccs          C31A A33T       1.000000   \n",
       "1    lib51  m64272e_230126_100200/143/ccs        C571G T572A       1.000000   \n",
       "2    lib51  m64272e_230126_100200/152/ccs        del134to136       1.000000   \n",
       "3    lib51  m64272e_230126_100200/184/ccs        C502A G504A       1.000000   \n",
       "4    lib51  m64272e_230126_100200/242/ccs        A118T C120T       0.999999   \n",
       "5    lib51  m64272e_230126_100200/277/ccs     T82A T83C C84T       1.000000   \n",
       "6    lib51  m64272e_230126_100200/326/ccs              A109G       1.000000   \n",
       "7    lib51  m64272e_230126_100200/335/ccs        T250C C251A       1.000000   \n",
       "8    lib51  m64272e_230126_100200/352/ccs  A190T G191T T192C       1.000000   \n",
       "9    lib51  m64272e_230126_100200/420/ccs  G235T A236T C237T       1.000000   \n",
       "10   lib51  m64272e_230126_100200/465/ccs              A488C       1.000000   \n",
       "11   lib51  m64272e_230126_100200/512/ccs     G43C G44A C45A       1.000000   \n",
       "\n",
       "             barcode  barcode_accuracy target  barcode_error    gene_error  \\\n",
       "0   TCAACTTTACCAGGTG               1.0   MERS   1.000000e-07  1.000000e-07   \n",
       "1   CTTATCCCTTAACAGT               1.0   MERS   1.000000e-07  1.000000e-07   \n",
       "2   CCATATCCCGCTCTTG               1.0   MERS   1.000000e-07  1.000000e-07   \n",
       "3   AACGGAAAAAAACCTT               1.0   MERS   1.000000e-07  1.000000e-07   \n",
       "4   ATGTTCGTCCCAGATA               1.0   MERS   1.000000e-07  5.087173e-07   \n",
       "5   AAGCAACTAAAAAACG               1.0   MERS   1.000000e-07  1.000000e-07   \n",
       "6   CCAGAGTTAAGACATT               1.0   MERS   1.000000e-07  1.000000e-07   \n",
       "7   ATTTACTAATCAATAA               1.0   MERS   1.000000e-07  1.000000e-07   \n",
       "8   CAAAGGATTCATAGCC               1.0   MERS   1.000000e-07  1.000000e-07   \n",
       "9   ATGGAGGGCATTAGGC               1.0   MERS   1.000000e-07  1.000000e-07   \n",
       "10  TCGCATGGGACTTAAG               1.0   MERS   1.000000e-07  1.000000e-07   \n",
       "11  ATAGATAAGAAATAAG               1.0   MERS   1.000000e-07  1.000000e-07   \n",
       "\n",
       "    retained gene_mutations_with_del  n_indels  has_indel  \n",
       "0       True               C31A A33T         0      False  \n",
       "1       True             C571G T572A         0      False  \n",
       "2       True       G136- A137- T138-         0      False  \n",
       "3       True             C502A G504A         0      False  \n",
       "4       True             A118T C120T         0      False  \n",
       "5       True          T82A T83C C84T         0      False  \n",
       "6       True                   A109G         0      False  \n",
       "7       True             T250C C251A         0      False  \n",
       "8       True       A190T G191T T192C         0      False  \n",
       "9       True       G235T A236T C237T         0      False  \n",
       "10      True                   A488C         0      False  \n",
       "11      True          G43C G44A C45A         0      False  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "processed_ccs = processed_ccs.reset_index(drop=True)\n",
    "\n",
    "deltosubs = alignparse.utils.InFrameDeletionsToSubs(geneseq)\n",
    "\n",
    "processed_ccs['gene_mutations_with_del'] = processed_ccs.apply(lambda x: deltosubs.dels_to_subs(x['gene_mutations']), axis=1)\n",
    "\n",
    "processed_ccs = alignparse.consensus.add_mut_info_cols(processed_ccs,\n",
    "                                                       mutation_col='gene_mutations_with_del',\n",
    "                                                       n_indel_col='n_indels')\n",
    "\n",
    "processed_ccs = processed_ccs.assign(has_indel=lambda x: x['n_indels'] > 0)\n",
    "\n",
    "processed_ccs.head(n=12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot how many sequences have indels:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = (\n",
    " ggplot(processed_ccs,\n",
    "        aes('retained', fill='has_indel')) +\n",
    " geom_bar(position='dodge') +\n",
    " geom_text(aes(label='..count..'), stat='count', va='bottom', size=7,\n",
    "           position=position_dodge(width=0.9), format_string='{:.2g}') +\n",
    " theme(figure_size=(2.5 * nlibs, 3),\n",
    "       panel_grid_major_x=element_blank(),\n",
    "       ) +\n",
    " ylab('number of CCSs') +\n",
    " scale_fill_manual(values=CBPALETTE[1:]) +\n",
    " facet_wrap('~ library', nrow=1)\n",
    " ).draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now get the empirical accuracy for each of the CCS groups mentioned above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_acc = config['max_error_rate'] / 10\n",
    "empirical_acc = []\n",
    "\n",
    "for desc, query_str in [\n",
    "        ('retained', 'retained'),\n",
    "        ('retained, no indel', 'retained and not has_indel'),\n",
    "        ('10X accuracy',\n",
    "         f\"(gene_error < {high_acc}) and (barcode_error < {high_acc})\"),\n",
    "        ('10X accuracy, no indel',\n",
    "         f\"(gene_error < {high_acc}) and (barcode_error < {high_acc}) and not has_indel\")\n",
    "        ]:\n",
    "    # get just CCSs in that category\n",
    "    df = processed_ccs.query(query_str)\n",
    "    \n",
    "    # compute empirical accuracy\n",
    "    empirical_acc.append(\n",
    "        alignparse.consensus.empirical_accuracy(df,\n",
    "                                                mutation_col='gene_mutations')\n",
    "        .assign(description=desc)\n",
    "        .merge(df\n",
    "               .groupby('library')\n",
    "               .size()\n",
    "               .rename('number_CCSs')\n",
    "               .reset_index()\n",
    "               )\n",
    "        )\n",
    "\n",
    "# make description categorical to preserve order, and annotate as \"actual\"\n",
    "# the category (\"retained, no indel\") that we will use for building variants.\n",
    "empirical_acc = (\n",
    "    pd.concat(empirical_acc, ignore_index=True, sort=False)\n",
    "    .assign(description=lambda x: pd.Categorical(x['description'],\n",
    "                                                 x['description'].unique(),\n",
    "                                                 ordered=True),\n",
    "            actual=lambda x: numpy.where(x['description'] == 'retained, no indel',\n",
    "                                         True, False),\n",
    "            )\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Display table of the empirical accuracies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>library</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>description</th>\n",
       "      <th>number_CCSs</th>\n",
       "      <th>actual</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>0.966829</td>\n",
       "      <td>retained</td>\n",
       "      <td>156851</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib52</td>\n",
       "      <td>0.964572</td>\n",
       "      <td>retained</td>\n",
       "      <td>167806</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>0.983186</td>\n",
       "      <td>retained, no indel</td>\n",
       "      <td>153394</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib52</td>\n",
       "      <td>0.980309</td>\n",
       "      <td>retained, no indel</td>\n",
       "      <td>164350</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>0.971903</td>\n",
       "      <td>10X accuracy</td>\n",
       "      <td>151185</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib52</td>\n",
       "      <td>0.970190</td>\n",
       "      <td>10X accuracy</td>\n",
       "      <td>161412</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>0.983160</td>\n",
       "      <td>10X accuracy, no indel</td>\n",
       "      <td>148648</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib52</td>\n",
       "      <td>0.980717</td>\n",
       "      <td>10X accuracy, no indel</td>\n",
       "      <td>158953</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(empirical_acc.to_html(index=False)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the empirical accuracies, using a different color to show the category that we will actually use:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving plot to results/figures/empirical_CCS_accuracy.pdf\n"
     ]
    }
   ],
   "source": [
    "p = (\n",
    "    ggplot(empirical_acc,\n",
    "           aes('description', 'accuracy', color='actual', label='accuracy')\n",
    "           ) +\n",
    "    geom_point(size=3) +\n",
    "    geom_text(va='bottom', size=9, format_string='{:.3g}', nudge_y=0.003) +\n",
    "    facet_wrap('~ library', ncol=8) +\n",
    "    theme(figure_size=(1.75 * nlibs, 2.25),\n",
    "          axis_text_x=element_text(angle=90),\n",
    "          panel_grid_major_x=element_blank(),\n",
    "          ) +\n",
    "    xlab('') +\n",
    "    scale_y_continuous(name='empirical accuracy', limits=(0.2, 1.005)) +\n",
    "    scale_color_manual(values=CBPALETTE, guide=False)\n",
    "    )\n",
    "\n",
    "plotfile = os.path.join(config['figs_dir'], 'empirical_CCS_accuracy.pdf')\n",
    "print(f\"Saving plot to {plotfile}\")\n",
    "_ = p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above analysis shows that if we exclude sequences with indels (which we plan to do among our consensus sequences), then the accuracy of each CCS is around 99%. \n",
    "We do **not** get notably higher empirical accuracy by imposing a more stringent filter from the PacBio `ccs` program, indicating that the major sources of error are due to processes that are not modeled in this program's accuracy filter (perhaps strand exchange or barcode sharing).\n",
    "\n",
    "Note that this empirical accuracy is for a **single** CCS.\n",
    "When we build the consensus sequences for each barcode below, we will take the consensus of CCSs within a barcode.\n",
    "So for barcodes with multiple CCSs, the actual accuracy of the consensus sequences will be higher than the empirical accuracy above due to capturing information from multiple CCSs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Consensus sequences for barcodes\n",
    "We call the consensus sequence for each barcode using the simple method implemented in [alignparse.consensus.simple_mutconsensus](https://jbloomlab.github.io/alignparse/alignparse.consensus.html?highlight=simple_mutconsensus#alignparse.consensus.simple_mutconsensus).\n",
    "The documentation for that function explains the method in detail, but basically it works like this:\n",
    " 1. When there is just one CCS per barcode, the consensus is just that sequence.\n",
    " 2. When there are multiple CCSs per barcode, they are used to build a consensus--however, the entire barcode is discarded if there are many differences between CCSs with the barcode, or high-frequency non-consensus mutations. The reason that barcodes are discarded in such cases as many differences between CCSs or high-frequency non-consensus mutations suggest errors such as barcode collisions or strand exchange.\n",
    " \n",
    "First, call the consensus for each barcode including **all** retained sequences, even those with undesirable indels."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus, dropped = alignparse.consensus.simple_mutconsensus(\n",
    "                        processed_ccs.query('retained'),\n",
    "                        group_cols=('library', 'barcode', 'target'),\n",
    "                        mutation_col='gene_mutations_with_del',\n",
    "                        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here are the first few lines of the data frame of consensus sequences for each barcode.\n",
    "In addition to giving the library, barcode, target, and mutations, it also has a column indicating how many CCSs support the variant call:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>library</th>\n",
       "      <th>barcode</th>\n",
       "      <th>target</th>\n",
       "      <th>gene_mutations_with_del</th>\n",
       "      <th>variant_call_support</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAACATTCGT</td>\n",
       "      <td>MERS</td>\n",
       "      <td>A433T A434T C435G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAAGACTTTC</td>\n",
       "      <td>MERS</td>\n",
       "      <td>G472A T473A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAAGCGATAG</td>\n",
       "      <td>MERS</td>\n",
       "      <td>T382- C383- C384-</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAATTGAGGT</td>\n",
       "      <td>MERS</td>\n",
       "      <td>C416A T417A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAACAATCCCG</td>\n",
       "      <td>MERS</td>\n",
       "      <td>A562T C563A A564T</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(consensus.head().to_html(index=False)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we retain variants with substitutions and single-codon deletions, add information about substitution mutations, and for troubleshooting, number of indels:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>library</th>\n",
       "      <th>barcode</th>\n",
       "      <th>target</th>\n",
       "      <th>gene_mutations_with_del</th>\n",
       "      <th>variant_call_support</th>\n",
       "      <th>substitutions</th>\n",
       "      <th>deletions</th>\n",
       "      <th>insertions</th>\n",
       "      <th>number_of_indels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAACATTCGT</td>\n",
       "      <td>MERS</td>\n",
       "      <td>A433T A434T C435G</td>\n",
       "      <td>1</td>\n",
       "      <td>A433T A434T C435G</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAAGACTTTC</td>\n",
       "      <td>MERS</td>\n",
       "      <td>G472A T473A</td>\n",
       "      <td>1</td>\n",
       "      <td>G472A T473A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAAGCGATAG</td>\n",
       "      <td>MERS</td>\n",
       "      <td>T382- C383- C384-</td>\n",
       "      <td>4</td>\n",
       "      <td>T382- C383- C384-</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAATTGAGGT</td>\n",
       "      <td>MERS</td>\n",
       "      <td>C416A T417A</td>\n",
       "      <td>2</td>\n",
       "      <td>C416A T417A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAACAATCCCG</td>\n",
       "      <td>MERS</td>\n",
       "      <td>A562T C563A A564T</td>\n",
       "      <td>4</td>\n",
       "      <td>A562T C563A A564T</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "consensus = alignparse.consensus.add_mut_info_cols(\n",
    "                    consensus,\n",
    "                    mutation_col='gene_mutations_with_del',\n",
    "                    sub_str_col='substitutions',\n",
    "                    del_str_col='deletions',\n",
    "                    ins_str_col='insertions',\n",
    "                    n_indel_col='number_of_indels',\n",
    "                    overwrite_cols=True)\n",
    "\n",
    "display(HTML(consensus.head().to_html(index=False)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot distribution of number of CCSs supporting each variant call (consensus), indicating whether or not there is an indel:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_variant_call_support = 6  # group variants with >= this much support\n",
    "\n",
    "_ = (\n",
    " ggplot(consensus\n",
    "        .assign(variant_call_support=lambda x: numpy.clip(x['variant_call_support'],\n",
    "                                                          None,\n",
    "                                                          max_variant_call_support),\n",
    "                indel_state=lambda x: numpy.where(x['number_of_indels'] > 0,\n",
    "                                                  'has indel', 'no indel')\n",
    "                ),\n",
    "        aes('variant_call_support')) +\n",
    " geom_bar() +\n",
    " ylab('number of variants') +\n",
    " facet_grid('indel_state ~ library') +\n",
    " theme(figure_size=(1.75 * nlibs, 3.5),\n",
    "       panel_grid_major_x=element_blank(),\n",
    "       ) \n",
    " ).draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that most variant consensus sequences do **not** have indels, especially if we limit to the more \"accurate\" ones that have multiple CCSs supporting them.\n",
    "\n",
    "We will ignore all consensus sequences with indels in the variant-barcode lookup table. \n",
    "We do this for two reasons:\n",
    " 1. When there is just one CCS supporting a consensus, it is less likely to be accurate as indels are the main mode of PacBio error.\n",
    " 2. For the purposes of our studies, we are interested in point mutations rather than indels anyway.\n",
    " \n",
    "Here are number of valid consensus sequence (no indels) for each library and target:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>library</th>\n",
       "      <th>target</th>\n",
       "      <th>consensus sequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>MERS</td>\n",
       "      <td>55135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib52</td>\n",
       "      <td>MERS</td>\n",
       "      <td>73735</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "consensus = consensus.query('number_of_indels < 1')\n",
    "\n",
    "lib_target_counts = (\n",
    "    consensus\n",
    "    .groupby(['library', 'target'])\n",
    "    .size()\n",
    "    .rename('consensus sequences')\n",
    "    .reset_index()\n",
    "    )\n",
    "\n",
    "display(HTML(lib_target_counts.to_html(index=False)))\n",
    "\n",
    "p = (ggplot(lib_target_counts.assign(xlabel=lambda x: x['target'] + ', ' + x['library']),\n",
    "            aes('xlabel', 'consensus sequences')) +\n",
    "     geom_point(size=3) +\n",
    "     theme(figure_size=(0.5 * nlibs * ntargets, 1.75),\n",
    "           axis_text_x=element_text(angle=90)) +\n",
    "     xlab('') +\n",
    "     scale_y_log10()\n",
    "     )\n",
    "\n",
    "_ = p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below we write the retained consensus sequences to a CSV file that links the nucleotide mutations to the barcodes.\n",
    "(The next section analyzes this variant table in detail, and provides have more precise information on the number of variants and relevant statistics):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing nucleotide variants to results/variants/nucleotide_variant_table_MERS.csv\n",
      "Here are the first few lines of this file:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>target</th>\n",
       "      <th>library</th>\n",
       "      <th>barcode</th>\n",
       "      <th>gene_mutations_with_del</th>\n",
       "      <th>substitutions</th>\n",
       "      <th>insertions</th>\n",
       "      <th>deletions</th>\n",
       "      <th>variant_call_support</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAACATTCGT</td>\n",
       "      <td>A433T A434T C435G</td>\n",
       "      <td>A433T A434T C435G</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAAGACTTTC</td>\n",
       "      <td>G472A T473A</td>\n",
       "      <td>G472A T473A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAAGCGATAG</td>\n",
       "      <td>T382- C383- C384-</td>\n",
       "      <td>T382- C383- C384-</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAATTGAGGT</td>\n",
       "      <td>C416A T417A</td>\n",
       "      <td>C416A T417A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAACAATCCCG</td>\n",
       "      <td>A562T C563A A564T</td>\n",
       "      <td>A562T C563A A564T</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(f\"Writing nucleotide variants to {config['nt_variant_table_file' + '_' + background]}\")\n",
    "      \n",
    "(consensus\n",
    " [['target', 'library', 'barcode', 'gene_mutations_with_del', 'substitutions', 'insertions','deletions', 'variant_call_support']]\n",
    " .to_csv(config['nt_variant_table_file' + '_' + background], index=False)\n",
    " )\n",
    "      \n",
    "print('Here are the first few lines of this file:')\n",
    "display(HTML(\n",
    "    pd.read_csv(config['nt_variant_table_file' + '_' + background], na_filter=None)\n",
    "    .head()\n",
    "    .to_html(index=False)\n",
    "    ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What happened to the barcodes that we \"dropped\" because we could not construct a reliable consensus?\n",
    "The `dropped` data frame from [alignparse.consensus.simple_mutconsensus](https://jbloomlab.github.io/alignparse/alignparse.consensus.html?highlight=simple_mutconsensus#alignparse.consensus.simple_mutconsensus) has this information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>library</th>\n",
       "      <th>barcode</th>\n",
       "      <th>target</th>\n",
       "      <th>drop_reason</th>\n",
       "      <th>nseqs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAACCCGATGAC</td>\n",
       "      <td>MERS</td>\n",
       "      <td>subs diff too large</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAACTAGAAAGCA</td>\n",
       "      <td>MERS</td>\n",
       "      <td>subs diff too large</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAACTAGCAGTAAT</td>\n",
       "      <td>MERS</td>\n",
       "      <td>subs diff too large</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAACTCTAAAATAT</td>\n",
       "      <td>MERS</td>\n",
       "      <td>subs diff too large</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAACTGGACGAAGA</td>\n",
       "      <td>MERS</td>\n",
       "      <td>subs diff too large</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(dropped.head().to_html(index=False)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Summarize the information in this data frame on dropped barcodes with the plot below.\n",
    "This plot shows several things.\n",
    "First, we see that the total number of barcodes dropped is modest (just a few thousand per library) relative to the total number of barcodes per library (seen above to be on the order of hundreds of thousands).\n",
    "Second, the main reason that barcodes are dropped is that there are CCSs within the same barcode with suspiciously large numbers of mutations relative to the consensus---which we use as a filter to discard the entire barcode as it could indicate strand exchange or some other issue.\n",
    "In any case, the modest number of dropped barcodes indicates that there probably isn't much of a need to worry: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_nseqs = 8  # plot together all barcodes with >= this many sequences\n",
    "\n",
    "_ = (\n",
    " ggplot(\n",
    "    dropped.assign(nseqs=lambda x: numpy.clip(x['nseqs'], None, max_nseqs)),\n",
    "    aes('nseqs')) + \n",
    " geom_bar() + \n",
    " scale_x_continuous(limits=(1, None)) +\n",
    " xlab('number of sequences for barcode') +\n",
    " ylab('number of barcodes') +\n",
    " facet_grid('library ~ drop_reason') +\n",
    " theme(figure_size=(10, 1.5 * nlibs),\n",
    "       panel_grid_major_x=element_blank(),\n",
    "       )\n",
    " ).draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create barcode-variant table\n",
    "We now create a [CodonVariantTable](https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable) that stores and processes all the information about the variant consensus sequences.\n",
    "Below we initialize such a table, and then analyze information about its composition. **Note, I believe this table eliminates our indel mutants, so we probably need to go back to the nt variants table to parse indels ourselves. Nonetheless, we'll keep this here so we can see what our (amino acid) mutant coverage looks like**\n",
    "\n",
    "### Initialize codon variant table\n",
    "In order to initialize the codon variant table, we need two pieces of information:\n",
    "  1. The wildtype gene sequence.\n",
    "  2. The list of nucleotide mutations for each variant as determined in the consensus calling above.\n",
    "\n",
    "Read \"wildtype\" gene sequence to which we made the alignments (in order to do this, initialize an `alignparse.Targets` and get the gene sequence from it):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read gene of 645 nts for MERS from data/PacBio_amplicon_MERS.gb\n"
     ]
    }
   ],
   "source": [
    "targets = alignparse.targets.Targets(seqsfile=config['amplicons_' + background],\n",
    "                                     feature_parse_specs=config['feature_parse_specs_' + background])\n",
    "geneseq = targets.get_target(background).get_feature('gene').seq\n",
    "\n",
    "print(f\"Read gene of {len(geneseq)} nts for {background} from {config['amplicons_' + background]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now initialize the codon variant table using this wildtype sequence and our list of nucleotide mutations for each variant:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>library</th>\n",
       "      <th>barcode</th>\n",
       "      <th>variant_call_support</th>\n",
       "      <th>codon_substitutions</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>n_codon_substitutions</th>\n",
       "      <th>n_aa_substitutions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAACATTCGT</td>\n",
       "      <td>1</td>\n",
       "      <td>AAC145TTG</td>\n",
       "      <td>N145L</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAAGACTTTC</td>\n",
       "      <td>1</td>\n",
       "      <td>GTA158AAA</td>\n",
       "      <td>V158K</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAAGCGATAG</td>\n",
       "      <td>4</td>\n",
       "      <td>TCC128---</td>\n",
       "      <td>S128-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAATTGAGGT</td>\n",
       "      <td>2</td>\n",
       "      <td>CCT139CAA</td>\n",
       "      <td>P139Q</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAACAATCCCG</td>\n",
       "      <td>4</td>\n",
       "      <td>ACA188TAT</td>\n",
       "      <td>T188Y</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAACCCATGGA</td>\n",
       "      <td>1</td>\n",
       "      <td>CCC54TGG</td>\n",
       "      <td>P54W</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAACTGTCAAT</td>\n",
       "      <td>5</td>\n",
       "      <td>GGG4GAA</td>\n",
       "      <td>G4E</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAACTTAAGAA</td>\n",
       "      <td>1</td>\n",
       "      <td>CTT191GTT</td>\n",
       "      <td>L191V</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAGCAATATA</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAGTAAGGAA</td>\n",
       "      <td>2</td>\n",
       "      <td>ATG76GAA</td>\n",
       "      <td>M76E</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAGTAGAATA</td>\n",
       "      <td>1</td>\n",
       "      <td>GGG176ACT</td>\n",
       "      <td>G176T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAGTTCATAT</td>\n",
       "      <td>2</td>\n",
       "      <td>CCG155TGT</td>\n",
       "      <td>P155C</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   target library           barcode  variant_call_support codon_substitutions  \\\n",
       "0    MERS   lib51  AAAAAAAAACATTCGT                     1           AAC145TTG   \n",
       "1    MERS   lib51  AAAAAAAAAGACTTTC                     1           GTA158AAA   \n",
       "2    MERS   lib51  AAAAAAAAAGCGATAG                     4           TCC128---   \n",
       "3    MERS   lib51  AAAAAAAAATTGAGGT                     2           CCT139CAA   \n",
       "4    MERS   lib51  AAAAAAAACAATCCCG                     4           ACA188TAT   \n",
       "5    MERS   lib51  AAAAAAAACCCATGGA                     1            CCC54TGG   \n",
       "6    MERS   lib51  AAAAAAAACTGTCAAT                     5             GGG4GAA   \n",
       "7    MERS   lib51  AAAAAAAACTTAAGAA                     1           CTT191GTT   \n",
       "8    MERS   lib51  AAAAAAAAGCAATATA                     2                       \n",
       "9    MERS   lib51  AAAAAAAAGTAAGGAA                     2            ATG76GAA   \n",
       "10   MERS   lib51  AAAAAAAAGTAGAATA                     1           GGG176ACT   \n",
       "11   MERS   lib51  AAAAAAAAGTTCATAT                     2           CCG155TGT   \n",
       "\n",
       "   aa_substitutions  n_codon_substitutions  n_aa_substitutions  \n",
       "0             N145L                      1                   1  \n",
       "1             V158K                      1                   1  \n",
       "2             S128-                      1                   1  \n",
       "3             P139Q                      1                   1  \n",
       "4             T188Y                      1                   1  \n",
       "5              P54W                      1                   1  \n",
       "6               G4E                      1                   1  \n",
       "7             L191V                      1                   1  \n",
       "8                                        0                   0  \n",
       "9              M76E                      1                   1  \n",
       "10            G176T                      1                   1  \n",
       "11            P155C                      1                   1  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants = dms_variants.codonvarianttable.CodonVariantTable(\n",
    "                barcode_variant_file=config['nt_variant_table_file_' + background],\n",
    "                geneseq=geneseq,\n",
    "                primary_target=background,\n",
    "                allowgaps=True,\n",
    "                )\n",
    "\n",
    "variants.barcode_variant_df.head(n=12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic stats on variants\n",
    "We now will analyze the variants.\n",
    "In this call and in the plots below, we set `samples=None` as we aren't looking at variant counts in specific samples, but are simply looking at properties of the variants in the table.\n",
    "\n",
    "Here are the number of variants for each target:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>library</th>\n",
       "      <th>lib51</th>\n",
       "      <th>lib52</th>\n",
       "      <th>all libraries</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>target</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MERS</th>\n",
       "      <td>55135</td>\n",
       "      <td>73735</td>\n",
       "      <td>128870</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(\n",
    "    variants\n",
    "    .n_variants_df(samples=None)\n",
    "    .pivot_table(index=['target'],\n",
    "                 columns='library',\n",
    "                 values='count')\n",
    "    .to_html()\n",
    "    ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the number of variants supported by each number of CCSs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_support = 10  # group variants with >= this much support\n",
    "\n",
    "p = variants.plotVariantSupportHistogram(max_support=max_support,\n",
    "                                         widthscale=1.1,\n",
    "                                         heightscale=0.9)\n",
    "p = p + theme(panel_grid_major_x=element_blank())  # no vertical grid lines\n",
    "_ = p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutations per variant\n",
    "Plot the number of barcoded variants with each number of amino-acid and codon mutations.\n",
    "This is for the primary target only, and doesn't include the spiked-in secondary targets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving plot to results/figures/n_aa_muts_per_variant_MERS.pdf\n",
      "Saving plot to results/figures/n_codon_muts_per_variant_MERS.pdf\n"
     ]
    }
   ],
   "source": [
    "max_muts = 7  # group all variants with >= this many mutations\n",
    "\n",
    "for mut_type in ['aa', 'codon']:\n",
    "    p = variants.plotNumMutsHistogram(mut_type, samples=None, max_muts=max_muts,\n",
    "                                      widthscale=1.1,\n",
    "                                      heightscale=0.9)\n",
    "    p = p + theme(panel_grid_major_x=element_blank())  # no vertical grid lines\n",
    "    _ = p.draw()\n",
    "    plotfile = os.path.join(config['figs_dir'], f\"n_{mut_type}_muts_per_variant_\"+background+\".pdf\")\n",
    "    print(f\"Saving plot to {plotfile}\")\n",
    "    p.save(plotfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the frequencies of different codon mutation types among **all** variants (any number of mutations), again only for primary target:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving plot to results/figures/avg_muts_per_variant_MERS.pdf\n"
     ]
    }
   ],
   "source": [
    "p = variants.plotNumCodonMutsByType(variant_type='all', samples=None,\n",
    "                                    ylabel='mutations per variant',\n",
    "                                    heightscale=0.8)\n",
    "p = p + theme(panel_grid_major_x=element_blank())  # no vertical grid lines\n",
    "_ = p.draw()\n",
    "plotfile = os.path.join(config['figs_dir'], f\"avg_muts_per_variant_\"+background+\".pdf\")\n",
    "print(f\"Saving plot to {plotfile}\")\n",
    "p.save(plotfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Variants supported by multiple PacBio CCSs should have fewer spurious mutations since sequencing errors are very unlikely to occur on two CCSs.\n",
    "Below we plot the number of codon mutations per variant among variants with at least two CCSs supporting their call.\n",
    "The difference in mutation rates here and in the plot above (that does not apply the `min_support=2` filter) gives some estimate of the frequency of mutations in our variants our spurious.\n",
    "In fact, we see the numbers are very similar, indicating that few of the mutations are spurious:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = variants.plotNumCodonMutsByType(variant_type='all', samples=None,\n",
    "                                    ylabel='mutations per variant', \n",
    "                                    min_support=2, heightscale=0.8)\n",
    "p = p + theme(panel_grid_major_x=element_blank())  # no vertical grid lines\n",
    "_ = p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Completeness of mutation sampling\n",
    "We examine how completely amino-acid mutations are sampled by the variants for the primary target, looking at single-mutant variants only and all variants.\n",
    "The plot below shows that virtually every mutation is found in a variant in each library, even if we just look among the single mutants.\n",
    "Things look especially good if we aggregate across libraries:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving plot to results/figures/variant_cumul_all_mut_coverage_MERS.pdf\n",
      "Saving plot to results/figures/variant_cumul_single_mut_coverage_MERS.pdf\n"
     ]
    }
   ],
   "source": [
    "for variant_type in ['all', 'single']:\n",
    "    p = variants.plotCumulMutCoverage(variant_type, mut_type='aa', samples=None)\n",
    "    _ = p.draw()\n",
    "    plotfile = os.path.join(config['figs_dir'],\n",
    "                            f\"variant_cumul_{variant_type}_mut_coverage_\"+background+\".pdf\")\n",
    "    print(f\"Saving plot to {plotfile}\")\n",
    "    p.save(plotfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get more quantitative information like that plotted above, we determine how many mutations are found 0, 1, or >1 times both among single and all mutants for the primary target:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>number_of_mutations</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>variant_type</th>\n",
       "      <th>include_stops</th>\n",
       "      <th>library</th>\n",
       "      <th>count</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"18\" valign=\"top\">all</th>\n",
       "      <th rowspan=\"9\" valign=\"top\">False</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">lib51</th>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;1</th>\n",
       "      <td>4288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">lib52</th>\n",
       "      <th>0</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;1</th>\n",
       "      <td>4293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">all libraries</th>\n",
       "      <th>0</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;1</th>\n",
       "      <td>4294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"9\" valign=\"top\">True</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">lib51</th>\n",
       "      <th>0</th>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;1</th>\n",
       "      <td>4346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">lib52</th>\n",
       "      <th>0</th>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;1</th>\n",
       "      <td>4355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">all libraries</th>\n",
       "      <th>0</th>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;1</th>\n",
       "      <td>4357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"18\" valign=\"top\">single</th>\n",
       "      <th rowspan=\"9\" valign=\"top\">False</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">lib51</th>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;1</th>\n",
       "      <td>4285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">lib52</th>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;1</th>\n",
       "      <td>4290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">all libraries</th>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;1</th>\n",
       "      <td>4293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"9\" valign=\"top\">True</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">lib51</th>\n",
       "      <th>0</th>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;1</th>\n",
       "      <td>4322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">lib52</th>\n",
       "      <th>0</th>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;1</th>\n",
       "      <td>4328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">all libraries</th>\n",
       "      <th>0</th>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;1</th>\n",
       "      <td>4331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "count_dfs = []\n",
    "for variant_type in ['all', 'single']:\n",
    "    i_counts = (variants.mutCounts(variant_type, mut_type='aa', samples=None)\n",
    "                .assign(variant_type=variant_type)\n",
    "                )\n",
    "    count_dfs += [i_counts.assign(include_stops=True),\n",
    "                  i_counts\n",
    "                  .query('not mutation.str.contains(\"\\*\")', engine='python')\n",
    "                  .assign(include_stops=False)\n",
    "                  ]\n",
    "    \n",
    "display(HTML(\n",
    "    pd.concat(count_dfs)\n",
    "    .assign(count=lambda x: (numpy.clip(x['count'], None, 2)\n",
    "                             .map({0: '0', 1: '1', 2:'>1'}))\n",
    "            )\n",
    "    .groupby(['variant_type', 'include_stops', 'library', 'count'])\n",
    "    .aggregate(number_of_mutations=pd.NamedAgg(column='mutation', aggfunc='count'))\n",
    "    .to_html()\n",
    "    ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutation frequencies along gene\n",
    "We plot the frequencies of mutations along the gene among the variants for the primary target.\n",
    "Ideally, this would be uniform.\n",
    "We make the plot for both all variants and single-mutant / wildtype variants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "for variant_type in ['all', 'single']:\n",
    "    p = variants.plotMutFreqs(variant_type, mut_type='codon', samples=None)\n",
    "    p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also use heat maps to examine the extent to which specific amino-acid or codon mutations are over-represented.\n",
    "These heat maps are large, so we make them just for all variants and the merge of all libraries:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "for mut_type in ['aa', 'codon']:\n",
    "    p = variants.plotMutHeatmap('all', mut_type, samples=None, #libraries='all_only',\n",
    "                                widthscale=2)\n",
    "    p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write codon-variant table\n",
    "We write the codon variant table to a CSV file.\n",
    "This table looks like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>target</th>\n",
       "      <th>library</th>\n",
       "      <th>barcode</th>\n",
       "      <th>variant_call_support</th>\n",
       "      <th>codon_substitutions</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>n_codon_substitutions</th>\n",
       "      <th>n_aa_substitutions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAACATTCGT</td>\n",
       "      <td>1</td>\n",
       "      <td>AAC145TTG</td>\n",
       "      <td>N145L</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAAGACTTTC</td>\n",
       "      <td>1</td>\n",
       "      <td>GTA158AAA</td>\n",
       "      <td>V158K</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAAGCGATAG</td>\n",
       "      <td>4</td>\n",
       "      <td>TCC128---</td>\n",
       "      <td>S128-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAAATTGAGGT</td>\n",
       "      <td>2</td>\n",
       "      <td>CCT139CAA</td>\n",
       "      <td>P139Q</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MERS</td>\n",
       "      <td>lib51</td>\n",
       "      <td>AAAAAAAACAATCCCG</td>\n",
       "      <td>4</td>\n",
       "      <td>ACA188TAT</td>\n",
       "      <td>T188Y</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(\n",
    "    variants.barcode_variant_df\n",
    "    .head()\n",
    "    .to_html(index=False)\n",
    "    ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note how this table differs from the nucleotide variant table we generated above and used to initialize the [CodonVariantTable](https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable) in that it gives **codon** substitutions and associated amino-acid substitutions.\n",
    "\n",
    "Write it to CSV file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing codon-variant table to results/variants/codon_variant_table_MERS.csv\n"
     ]
    }
   ],
   "source": [
    "print(f\"Writing codon-variant table to {config['codon_variant_table_file_'+background]}\")\n",
    "\n",
    "variants.barcode_variant_df.to_csv(config['codon_variant_table_file_'+background], index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
